// primeng/bundles/primeng-treetable.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('rxjs'), require('primeng/dom'), require('primeng/paginator'), require('primeng/api'), require('primeng/utils'), require('@angular/cdk/scrolling')) :
    typeof define === 'function' && define.amd ? define('primeng/treetable', ['exports', '@angular/core', '@angular/common', 'rxjs', 'primeng/dom', 'primeng/paginator', 'primeng/api', 'primeng/utils', '@angular/cdk/scrolling'], factory) :
    (global = global || self, factory((global.primeng = global.primeng || {}, global.primeng.treetable = {}), global.ng.core, global.ng.common, global.rxjs, global.primeng.dom, global.primeng.paginator, global.primeng.api, global.primeng.utils, global.ng.cdk.scrolling));
}(this, (function (exports, core, common, rxjs, dom, paginator, api, utils, scrolling) { 'use strict';

    var __assign = (this && this.__assign) || function () {
        __assign = Object.assign || function(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __values = (this && this.__values) || function(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    };
    var __read = (this && this.__read) || function (o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    };
    var __spread = (this && this.__spread) || function () {
        for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
        return ar;
    };
    var TreeTableService = /** @class */ (function () {
        function TreeTableService() {
            this.sortSource = new rxjs.Subject();
            this.selectionSource = new rxjs.Subject();
            this.contextMenuSource = new rxjs.Subject();
            this.uiUpdateSource = new rxjs.Subject();
            this.totalRecordsSource = new rxjs.Subject();
            this.sortSource$ = this.sortSource.asObservable();
            this.selectionSource$ = this.selectionSource.asObservable();
            this.contextMenuSource$ = this.contextMenuSource.asObservable();
            this.uiUpdateSource$ = this.uiUpdateSource.asObservable();
            this.totalRecordsSource$ = this.totalRecordsSource.asObservable();
        }
        TreeTableService.prototype.onSort = function (sortMeta) {
            this.sortSource.next(sortMeta);
        };
        TreeTableService.prototype.onSelectionChange = function () {
            this.selectionSource.next();
        };
        TreeTableService.prototype.onContextMenu = function (node) {
            this.contextMenuSource.next(node);
        };
        TreeTableService.prototype.onUIUpdate = function (value) {
            this.uiUpdateSource.next(value);
        };
        TreeTableService.prototype.onTotalRecordsChange = function (value) {
            this.totalRecordsSource.next(value);
        };
        TreeTableService = __decorate([
            core.Injectable()
        ], TreeTableService);
        return TreeTableService;
    }());
    var TreeTable = /** @class */ (function () {
        function TreeTable(el, zone, tableService) {
            this.el = el;
            this.zone = zone;
            this.tableService = tableService;
            this.lazy = false;
            this.lazyLoadOnInit = true;
            this.first = 0;
            this.pageLinks = 5;
            this.alwaysShowPaginator = true;
            this.paginatorPosition = 'bottom';
            this.currentPageReportTemplate = '{currentPage} of {totalPages}';
            this.defaultSortOrder = 1;
            this.sortMode = 'single';
            this.resetPageOnSort = true;
            this.selectionChange = new core.EventEmitter();
            this.contextMenuSelectionChange = new core.EventEmitter();
            this.contextMenuSelectionMode = "separate";
            this.compareSelectionBy = 'deepEquals';
            this.loadingIcon = 'pi pi-spinner';
            this.showLoader = true;
            this.virtualScrollDelay = 150;
            this.virtualRowHeight = 28;
            this.columnResizeMode = 'fit';
            this.rowTrackBy = function (index, item) { return item; };
            this.filters = {};
            this.filterDelay = 300;
            this.filterMode = 'lenient';
            this.onFilter = new core.EventEmitter();
            this.onNodeExpand = new core.EventEmitter();
            this.onNodeCollapse = new core.EventEmitter();
            this.onPage = new core.EventEmitter();
            this.onSort = new core.EventEmitter();
            this.onLazyLoad = new core.EventEmitter();
            this.sortFunction = new core.EventEmitter();
            this.onColResize = new core.EventEmitter();
            this.onColReorder = new core.EventEmitter();
            this.onNodeSelect = new core.EventEmitter();
            this.onNodeUnselect = new core.EventEmitter();
            this.onContextMenuSelect = new core.EventEmitter();
            this.onHeaderCheckboxToggle = new core.EventEmitter();
            this.onEditInit = new core.EventEmitter();
            this.onEditComplete = new core.EventEmitter();
            this.onEditCancel = new core.EventEmitter();
            this._value = [];
            this._totalRecords = 0;
            this._sortOrder = 1;
            this.selectionKeys = {};
        }
        TreeTable.prototype.ngOnInit = function () {
            if (this.lazy && this.lazyLoadOnInit) {
                this.onLazyLoad.emit(this.createLazyLoadMetadata());
            }
            this.initialized = true;
        };
        TreeTable.prototype.ngAfterContentInit = function () {
            var _this = this;
            this.templates.forEach(function (item) {
                switch (item.getType()) {
                    case 'caption':
                        _this.captionTemplate = item.template;
                        break;
                    case 'header':
                        _this.headerTemplate = item.template;
                        break;
                    case 'body':
                        _this.bodyTemplate = item.template;
                        break;
                    case 'loadingbody':
                        _this.loadingBodyTemplate = item.template;
                        break;
                    case 'footer':
                        _this.footerTemplate = item.template;
                        break;
                    case 'summary':
                        _this.summaryTemplate = item.template;
                        break;
                    case 'colgroup':
                        _this.colGroupTemplate = item.template;
                        break;
                    case 'emptymessage':
                        _this.emptyMessageTemplate = item.template;
                        break;
                    case 'paginatorleft':
                        _this.paginatorLeftTemplate = item.template;
                        break;
                    case 'paginatorright':
                        _this.paginatorRightTemplate = item.template;
                        break;
                    case 'frozenheader':
                        _this.frozenHeaderTemplate = item.template;
                        break;
                    case 'frozenbody':
                        _this.frozenBodyTemplate = item.template;
                        break;
                    case 'frozenfooter':
                        _this.frozenFooterTemplate = item.template;
                        break;
                    case 'frozencolgroup':
                        _this.frozenColGroupTemplate = item.template;
                        break;
                }
            });
        };
        TreeTable.prototype.ngOnChanges = function (simpleChange) {
            if (simpleChange.value) {
                this._value = simpleChange.value.currentValue;
                if (!this.lazy) {
                    this.totalRecords = (this._value ? this._value.length : 0);
                    if (this.sortMode == 'single' && this.sortField)
                        this.sortSingle();
                    else if (this.sortMode == 'multiple' && this.multiSortMeta)
                        this.sortMultiple();
                    else if (this.hasFilter()) //sort already filters
                        this._filter();
                }
                this.updateSerializedValue();
                this.tableService.onUIUpdate(this.value);
            }
            if (simpleChange.sortField) {
                this._sortField = simpleChange.sortField.currentValue;
                //avoid triggering lazy load prior to lazy initialization at onInit
                if (!this.lazy || this.initialized) {
                    if (this.sortMode === 'single') {
                        this.sortSingle();
                    }
                }
            }
            if (simpleChange.sortOrder) {
                this._sortOrder = simpleChange.sortOrder.currentValue;
                //avoid triggering lazy load prior to lazy initialization at onInit
                if (!this.lazy || this.initialized) {
                    if (this.sortMode === 'single') {
                        this.sortSingle();
                    }
                }
            }
            if (simpleChange.multiSortMeta) {
                this._multiSortMeta = simpleChange.multiSortMeta.currentValue;
                if (this.sortMode === 'multiple') {
                    this.sortMultiple();
                }
            }
            if (simpleChange.selection) {
                this._selection = simpleChange.selection.currentValue;
                if (!this.preventSelectionSetterPropagation) {
                    this.updateSelectionKeys();
                    this.tableService.onSelectionChange();
                }
                this.preventSelectionSetterPropagation = false;
            }
        };
        Object.defineProperty(TreeTable.prototype, "value", {
            get: function () {
                return this._value;
            },
            set: function (val) {
                this._value = val;
            },
            enumerable: true,
            configurable: true
        });
        TreeTable.prototype.updateSerializedValue = function () {
            this.serializedValue = [];
            if (this.paginator)
                this.serializePageNodes();
            else
                this.serializeNodes(null, this.filteredNodes || this.value, 0, true);
        };
        TreeTable.prototype.serializeNodes = function (parent, nodes, level, visible) {
            var e_1, _a;
            if (nodes && nodes.length) {
                try {
                    for (var nodes_1 = __values(nodes), nodes_1_1 = nodes_1.next(); !nodes_1_1.done; nodes_1_1 = nodes_1.next()) {
                        var node = nodes_1_1.value;
                        node.parent = parent;
                        var rowNode = {
                            node: node,
                            parent: parent,
                            level: level,
                            visible: visible && (parent ? parent.expanded : true)
                        };
                        this.serializedValue.push(rowNode);
                        if (rowNode.visible && node.expanded) {
                            this.serializeNodes(node, node.children, level + 1, rowNode.visible);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (nodes_1_1 && !nodes_1_1.done && (_a = nodes_1.return)) _a.call(nodes_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
        };
        TreeTable.prototype.serializePageNodes = function () {
            var data = this.filteredNodes || this.value;
            this.serializedValue = [];
            if (data && data.length) {
                var first = this.lazy ? 0 : this.first;
                for (var i = first; i < (first + this.rows); i++) {
                    var node = data[i];
                    if (node) {
                        this.serializedValue.push({
                            node: node,
                            parent: null,
                            level: 0,
                            visible: true
                        });
                        this.serializeNodes(node, node.children, 1, true);
                    }
                }
            }
        };
        Object.defineProperty(TreeTable.prototype, "totalRecords", {
            get: function () {
                return this._totalRecords;
            },
            set: function (val) {
                this._totalRecords = val;
                this.tableService.onTotalRecordsChange(this._totalRecords);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TreeTable.prototype, "sortField", {
            get: function () {
                return this._sortField;
            },
            set: function (val) {
                this._sortField = val;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TreeTable.prototype, "sortOrder", {
            get: function () {
                return this._sortOrder;
            },
            set: function (val) {
                this._sortOrder = val;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TreeTable.prototype, "multiSortMeta", {
            get: function () {
                return this._multiSortMeta;
            },
            set: function (val) {
                this._multiSortMeta = val;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TreeTable.prototype, "selection", {
            get: function () {
                return this._selection;
            },
            set: function (val) {
                this._selection = val;
            },
            enumerable: true,
            configurable: true
        });
        TreeTable.prototype.updateSelectionKeys = function () {
            var e_2, _a;
            if (this.dataKey && this._selection) {
                this.selectionKeys = {};
                if (Array.isArray(this._selection)) {
                    try {
                        for (var _b = __values(this._selection), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var node = _c.value;
                            this.selectionKeys[String(utils.ObjectUtils.resolveFieldData(node.data, this.dataKey))] = 1;
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                }
                else {
                    this.selectionKeys[String(utils.ObjectUtils.resolveFieldData(this._selection.data, this.dataKey))] = 1;
                }
            }
        };
        TreeTable.prototype.onPageChange = function (event) {
            this.first = event.first;
            this.rows = event.rows;
            if (this.lazy)
                this.onLazyLoad.emit(this.createLazyLoadMetadata());
            else
                this.serializePageNodes();
            this.onPage.emit({
                first: this.first,
                rows: this.rows
            });
            this.tableService.onUIUpdate(this.value);
            if (this.scrollable) {
                this.resetScrollTop();
            }
        };
        TreeTable.prototype.sort = function (event) {
            var originalEvent = event.originalEvent;
            if (this.sortMode === 'single') {
                this._sortOrder = (this.sortField === event.field) ? this.sortOrder * -1 : this.defaultSortOrder;
                this._sortField = event.field;
                this.sortSingle();
                if (this.resetPageOnSort && this.scrollable) {
                    this.resetScrollTop();
                }
            }
            if (this.sortMode === 'multiple') {
                var metaKey = originalEvent.metaKey || originalEvent.ctrlKey;
                var sortMeta = this.getSortMeta(event.field);
                if (sortMeta) {
                    if (!metaKey) {
                        this._multiSortMeta = [{ field: event.field, order: sortMeta.order * -1 }];
                        if (this.resetPageOnSort && this.scrollable) {
                            this.resetScrollTop();
                        }
                    }
                    else {
                        sortMeta.order = sortMeta.order * -1;
                    }
                }
                else {
                    if (!metaKey || !this.multiSortMeta) {
                        this._multiSortMeta = [];
                        if (this.resetPageOnSort && this.scrollable) {
                            this.resetScrollTop();
                        }
                    }
                    this.multiSortMeta.push({ field: event.field, order: this.defaultSortOrder });
                }
                this.sortMultiple();
            }
        };
        TreeTable.prototype.sortSingle = function () {
            if (this.sortField && this.sortOrder) {
                if (this.lazy) {
                    this.onLazyLoad.emit(this.createLazyLoadMetadata());
                }
                else if (this.value) {
                    this.sortNodes(this.value);
                    if (this.hasFilter()) {
                        this._filter();
                    }
                }
                var sortMeta = {
                    field: this.sortField,
                    order: this.sortOrder
                };
                this.onSort.emit(sortMeta);
                this.tableService.onSort(sortMeta);
                this.updateSerializedValue();
            }
        };
        TreeTable.prototype.sortNodes = function (nodes) {
            var e_3, _a;
            var _this = this;
            if (!nodes || nodes.length === 0) {
                return;
            }
            if (this.customSort) {
                this.sortFunction.emit({
                    data: nodes,
                    mode: this.sortMode,
                    field: this.sortField,
                    order: this.sortOrder
                });
            }
            else {
                nodes.sort(function (node1, node2) {
                    var value1 = utils.ObjectUtils.resolveFieldData(node1.data, _this.sortField);
                    var value2 = utils.ObjectUtils.resolveFieldData(node2.data, _this.sortField);
                    var result = null;
                    if (value1 == null && value2 != null)
                        result = -1;
                    else if (value1 != null && value2 == null)
                        result = 1;
                    else if (value1 == null && value2 == null)
                        result = 0;
                    else if (typeof value1 === 'string' && typeof value2 === 'string')
                        result = value1.localeCompare(value2, undefined, { numeric: true });
                    else
                        result = (value1 < value2) ? -1 : (value1 > value2) ? 1 : 0;
                    return (_this.sortOrder * result);
                });
            }
            try {
                for (var nodes_2 = __values(nodes), nodes_2_1 = nodes_2.next(); !nodes_2_1.done; nodes_2_1 = nodes_2.next()) {
                    var node = nodes_2_1.value;
                    this.sortNodes(node.children);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (nodes_2_1 && !nodes_2_1.done && (_a = nodes_2.return)) _a.call(nodes_2);
                }
                finally { if (e_3) throw e_3.error; }
            }
        };
        TreeTable.prototype.sortMultiple = function () {
            if (this.multiSortMeta) {
                if (this.lazy) {
                    this.onLazyLoad.emit(this.createLazyLoadMetadata());
                }
                else if (this.value) {
                    this.sortMultipleNodes(this.value);
                    if (this.hasFilter()) {
                        this._filter();
                    }
                }
                this.onSort.emit({
                    multisortmeta: this.multiSortMeta
                });
                this.tableService.onSort(this.multiSortMeta);
                this.updateSerializedValue();
            }
        };
        TreeTable.prototype.sortMultipleNodes = function (nodes) {
            var e_4, _a;
            var _this = this;
            if (!nodes || nodes.length === 0) {
                return;
            }
            if (this.customSort) {
                this.sortFunction.emit({
                    data: this.value,
                    mode: this.sortMode,
                    multiSortMeta: this.multiSortMeta
                });
            }
            else {
                nodes.sort(function (node1, node2) {
                    return _this.multisortField(node1, node2, _this.multiSortMeta, 0);
                });
            }
            try {
                for (var nodes_3 = __values(nodes), nodes_3_1 = nodes_3.next(); !nodes_3_1.done; nodes_3_1 = nodes_3.next()) {
                    var node = nodes_3_1.value;
                    this.sortMultipleNodes(node.children);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (nodes_3_1 && !nodes_3_1.done && (_a = nodes_3.return)) _a.call(nodes_3);
                }
                finally { if (e_4) throw e_4.error; }
            }
        };
        TreeTable.prototype.multisortField = function (node1, node2, multiSortMeta, index) {
            var value1 = utils.ObjectUtils.resolveFieldData(node1.data, multiSortMeta[index].field);
            var value2 = utils.ObjectUtils.resolveFieldData(node2.data, multiSortMeta[index].field);
            var result = null;
            if (value1 == null && value2 != null)
                result = -1;
            else if (value1 != null && value2 == null)
                result = 1;
            else if (value1 == null && value2 == null)
                result = 0;
            if (typeof value1 == 'string' || value1 instanceof String) {
                if (value1.localeCompare && (value1 != value2)) {
                    return (multiSortMeta[index].order * value1.localeCompare(value2, undefined, { numeric: true }));
                }
            }
            else {
                result = (value1 < value2) ? -1 : 1;
            }
            if (value1 == value2) {
                return (multiSortMeta.length - 1) > (index) ? (this.multisortField(node1, node2, multiSortMeta, index + 1)) : 0;
            }
            return (multiSortMeta[index].order * result);
        };
        TreeTable.prototype.getSortMeta = function (field) {
            if (this.multiSortMeta && this.multiSortMeta.length) {
                for (var i = 0; i < this.multiSortMeta.length; i++) {
                    if (this.multiSortMeta[i].field === field) {
                        return this.multiSortMeta[i];
                    }
                }
            }
            return null;
        };
        TreeTable.prototype.isSorted = function (field) {
            if (this.sortMode === 'single') {
                return (this.sortField && this.sortField === field);
            }
            else if (this.sortMode === 'multiple') {
                var sorted = false;
                if (this.multiSortMeta) {
                    for (var i = 0; i < this.multiSortMeta.length; i++) {
                        if (this.multiSortMeta[i].field == field) {
                            sorted = true;
                            break;
                        }
                    }
                }
                return sorted;
            }
        };
        TreeTable.prototype.createLazyLoadMetadata = function () {
            return {
                first: this.first,
                rows: this.rows,
                sortField: this.sortField,
                sortOrder: this.sortOrder,
                filters: this.filters,
                globalFilter: this.filters && this.filters['global'] ? this.filters['global'].value : null,
                multiSortMeta: this.multiSortMeta
            };
        };
        TreeTable.prototype.resetScrollTop = function () {
            if (this.virtualScroll)
                this.scrollToVirtualIndex(0);
            else
                this.scrollTo({ top: 0 });
        };
        TreeTable.prototype.scrollToVirtualIndex = function (index) {
            if (this.scrollableViewChild) {
                this.scrollableViewChild.scrollToVirtualIndex(index);
            }
            if (this.scrollableFrozenViewChild) {
                this.scrollableFrozenViewChild.scrollToVirtualIndex(index);
            }
        };
        TreeTable.prototype.scrollTo = function (options) {
            if (this.scrollableViewChild) {
                this.scrollableViewChild.scrollTo(options);
            }
            if (this.scrollableFrozenViewChild) {
                this.scrollableFrozenViewChild.scrollTo(options);
            }
        };
        TreeTable.prototype.isEmpty = function () {
            var data = this.filteredNodes || this.value;
            return data == null || data.length == 0;
        };
        TreeTable.prototype.getBlockableElement = function () {
            return this.el.nativeElement.children[0];
        };
        TreeTable.prototype.onColumnResizeBegin = function (event) {
            var containerLeft = dom.DomHandler.getOffset(this.containerViewChild.nativeElement).left;
            this.lastResizerHelperX = (event.pageX - containerLeft + this.containerViewChild.nativeElement.scrollLeft);
            event.preventDefault();
        };
        TreeTable.prototype.onColumnResize = function (event) {
            var containerLeft = dom.DomHandler.getOffset(this.containerViewChild.nativeElement).left;
            dom.DomHandler.addClass(this.containerViewChild.nativeElement, 'ui-unselectable-text');
            this.resizeHelperViewChild.nativeElement.style.height = this.containerViewChild.nativeElement.offsetHeight + 'px';
            this.resizeHelperViewChild.nativeElement.style.top = 0 + 'px';
            this.resizeHelperViewChild.nativeElement.style.left = (event.pageX - containerLeft + this.containerViewChild.nativeElement.scrollLeft) + 'px';
            this.resizeHelperViewChild.nativeElement.style.display = 'block';
        };
        TreeTable.prototype.onColumnResizeEnd = function (event, column) {
            var delta = this.resizeHelperViewChild.nativeElement.offsetLeft - this.lastResizerHelperX;
            var columnWidth = column.offsetWidth;
            var newColumnWidth = columnWidth + delta;
            var minWidth = column.style.minWidth || 15;
            if (columnWidth + delta > parseInt(minWidth)) {
                if (this.columnResizeMode === 'fit') {
                    var nextColumn = column.nextElementSibling;
                    while (!nextColumn.offsetParent) {
                        nextColumn = nextColumn.nextElementSibling;
                    }
                    if (nextColumn) {
                        var nextColumnWidth = nextColumn.offsetWidth - delta;
                        var nextColumnMinWidth = nextColumn.style.minWidth || 15;
                        if (newColumnWidth > 15 && nextColumnWidth > parseInt(nextColumnMinWidth)) {
                            if (this.scrollable) {
                                var scrollableView = this.findParentScrollableView(column);
                                var scrollableBodyTable = dom.DomHandler.findSingle(scrollableView, '.ui-treetable-scrollable-body table');
                                var scrollableHeaderTable = dom.DomHandler.findSingle(scrollableView, 'table.ui-treetable-scrollable-header-table');
                                var scrollableFooterTable = dom.DomHandler.findSingle(scrollableView, 'table.ui-treetable-scrollable-footer-table');
                                var resizeColumnIndex = dom.DomHandler.index(column);
                                this.resizeColGroup(scrollableHeaderTable, resizeColumnIndex, newColumnWidth, nextColumnWidth);
                                this.resizeColGroup(scrollableBodyTable, resizeColumnIndex, newColumnWidth, nextColumnWidth);
                                this.resizeColGroup(scrollableFooterTable, resizeColumnIndex, newColumnWidth, nextColumnWidth);
                            }
                            else {
                                column.style.width = newColumnWidth + 'px';
                                if (nextColumn) {
                                    nextColumn.style.width = nextColumnWidth + 'px';
                                }
                            }
                        }
                    }
                }
                else if (this.columnResizeMode === 'expand') {
                    if (this.scrollable) {
                        var scrollableView = this.findParentScrollableView(column);
                        var scrollableBody_1 = dom.DomHandler.findSingle(scrollableView, '.ui-treetable-scrollable-body');
                        var scrollableBodyTable = dom.DomHandler.findSingle(scrollableView, '.ui-treetable-scrollable-body table');
                        var scrollableHeader = dom.DomHandler.findSingle(scrollableView, '.ui-treetable-scrollable-header');
                        var scrollableFooter = dom.DomHandler.findSingle(scrollableView, '.ui-treetable-scrollable-footer');
                        var scrollableHeaderTable = dom.DomHandler.findSingle(scrollableView, 'table.ui-treetable-scrollable-header-table');
                        var scrollableFooterTable = dom.DomHandler.findSingle(scrollableView, 'table.ui-treetable-scrollable-footer-table');
                        scrollableBodyTable.style.width = scrollableBodyTable.offsetWidth + delta + 'px';
                        scrollableHeaderTable.style.width = scrollableHeaderTable.offsetWidth + delta + 'px';
                        if (scrollableFooterTable) {
                            scrollableFooterTable.style.width = scrollableFooterTable.offsetWidth + delta + 'px';
                        }
                        var resizeColumnIndex = dom.DomHandler.index(column);
                        var scrollableBodyTableWidth = column ? scrollableBodyTable.offsetWidth + delta : newColumnWidth;
                        var scrollableHeaderTableWidth = column ? scrollableHeaderTable.offsetWidth + delta : newColumnWidth;
                        var isContainerInViewport = this.containerViewChild.nativeElement.offsetWidth >= scrollableBodyTableWidth;
                        var setWidth = function (container, table, width, isContainerInViewport) {
                            if (container && table) {
                                container.style.width = isContainerInViewport ? width + dom.DomHandler.calculateScrollbarWidth(scrollableBody_1) + 'px' : 'auto';
                                table.style.width = width + 'px';
                            }
                        };
                        setWidth(scrollableBody_1, scrollableBodyTable, scrollableBodyTableWidth, isContainerInViewport);
                        setWidth(scrollableHeader, scrollableHeaderTable, scrollableHeaderTableWidth, isContainerInViewport);
                        setWidth(scrollableFooter, scrollableFooterTable, scrollableHeaderTableWidth, isContainerInViewport);
                        this.resizeColGroup(scrollableHeaderTable, resizeColumnIndex, newColumnWidth, null);
                        this.resizeColGroup(scrollableBodyTable, resizeColumnIndex, newColumnWidth, null);
                        this.resizeColGroup(scrollableFooterTable, resizeColumnIndex, newColumnWidth, null);
                    }
                    else {
                        this.tableViewChild.nativeElement.style.width = this.tableViewChild.nativeElement.offsetWidth + delta + 'px';
                        column.style.width = newColumnWidth + 'px';
                        var containerWidth = this.tableViewChild.nativeElement.style.width;
                        this.containerViewChild.nativeElement.style.width = containerWidth + 'px';
                    }
                }
                this.onColResize.emit({
                    element: column,
                    delta: delta
                });
            }
            this.resizeHelperViewChild.nativeElement.style.display = 'none';
            dom.DomHandler.removeClass(this.containerViewChild.nativeElement, 'ui-unselectable-text');
        };
        TreeTable.prototype.findParentScrollableView = function (column) {
            if (column) {
                var parent_1 = column.parentElement;
                while (parent_1 && !dom.DomHandler.hasClass(parent_1, 'ui-treetable-scrollable-view')) {
                    parent_1 = parent_1.parentElement;
                }
                return parent_1;
            }
            else {
                return null;
            }
        };
        TreeTable.prototype.resizeColGroup = function (table, resizeColumnIndex, newColumnWidth, nextColumnWidth) {
            if (table) {
                var colGroup = table.children[0].nodeName === 'COLGROUP' ? table.children[0] : null;
                if (colGroup) {
                    var col = colGroup.children[resizeColumnIndex];
                    var nextCol = col.nextElementSibling;
                    col.style.width = newColumnWidth + 'px';
                    if (nextCol && nextColumnWidth) {
                        nextCol.style.width = nextColumnWidth + 'px';
                    }
                }
                else {
                    throw "Scrollable tables require a colgroup to support resizable columns";
                }
            }
        };
        TreeTable.prototype.onColumnDragStart = function (event, columnElement) {
            this.reorderIconWidth = dom.DomHandler.getHiddenElementOuterWidth(this.reorderIndicatorUpViewChild.nativeElement);
            this.reorderIconHeight = dom.DomHandler.getHiddenElementOuterHeight(this.reorderIndicatorDownViewChild.nativeElement);
            this.draggedColumn = columnElement;
            event.dataTransfer.setData('text', 'b'); // For firefox
        };
        TreeTable.prototype.onColumnDragEnter = function (event, dropHeader) {
            if (this.reorderableColumns && this.draggedColumn && dropHeader) {
                event.preventDefault();
                var containerOffset = dom.DomHandler.getOffset(this.containerViewChild.nativeElement);
                var dropHeaderOffset = dom.DomHandler.getOffset(dropHeader);
                if (this.draggedColumn != dropHeader) {
                    var targetLeft = dropHeaderOffset.left - containerOffset.left;
                    var targetTop = containerOffset.top - dropHeaderOffset.top;
                    var columnCenter = dropHeaderOffset.left + dropHeader.offsetWidth / 2;
                    this.reorderIndicatorUpViewChild.nativeElement.style.top = dropHeaderOffset.top - containerOffset.top - (this.reorderIconHeight - 1) + 'px';
                    this.reorderIndicatorDownViewChild.nativeElement.style.top = dropHeaderOffset.top - containerOffset.top + dropHeader.offsetHeight + 'px';
                    if (event.pageX > columnCenter) {
                        this.reorderIndicatorUpViewChild.nativeElement.style.left = (targetLeft + dropHeader.offsetWidth - Math.ceil(this.reorderIconWidth / 2)) + 'px';
                        this.reorderIndicatorDownViewChild.nativeElement.style.left = (targetLeft + dropHeader.offsetWidth - Math.ceil(this.reorderIconWidth / 2)) + 'px';
                        this.dropPosition = 1;
                    }
                    else {
                        this.reorderIndicatorUpViewChild.nativeElement.style.left = (targetLeft - Math.ceil(this.reorderIconWidth / 2)) + 'px';
                        this.reorderIndicatorDownViewChild.nativeElement.style.left = (targetLeft - Math.ceil(this.reorderIconWidth / 2)) + 'px';
                        this.dropPosition = -1;
                    }
                    this.reorderIndicatorUpViewChild.nativeElement.style.display = 'block';
                    this.reorderIndicatorDownViewChild.nativeElement.style.display = 'block';
                }
                else {
                    event.dataTransfer.dropEffect = 'none';
                }
            }
        };
        TreeTable.prototype.onColumnDragLeave = function (event) {
            if (this.reorderableColumns && this.draggedColumn) {
                event.preventDefault();
                this.reorderIndicatorUpViewChild.nativeElement.style.display = 'none';
                this.reorderIndicatorDownViewChild.nativeElement.style.display = 'none';
            }
        };
        TreeTable.prototype.onColumnDrop = function (event, dropColumn) {
            event.preventDefault();
            if (this.draggedColumn) {
                var dragIndex = dom.DomHandler.indexWithinGroup(this.draggedColumn, 'ttreorderablecolumn');
                var dropIndex = dom.DomHandler.indexWithinGroup(dropColumn, 'ttreorderablecolumn');
                var allowDrop = (dragIndex != dropIndex);
                if (allowDrop && ((dropIndex - dragIndex == 1 && this.dropPosition === -1) || (dragIndex - dropIndex == 1 && this.dropPosition === 1))) {
                    allowDrop = false;
                }
                if (allowDrop && ((dropIndex < dragIndex && this.dropPosition === 1))) {
                    dropIndex = dropIndex + 1;
                }
                if (allowDrop && ((dropIndex > dragIndex && this.dropPosition === -1))) {
                    dropIndex = dropIndex - 1;
                }
                if (allowDrop) {
                    utils.ObjectUtils.reorderArray(this.columns, dragIndex, dropIndex);
                    this.onColReorder.emit({
                        dragIndex: dragIndex,
                        dropIndex: dropIndex,
                        columns: this.columns
                    });
                }
                this.reorderIndicatorUpViewChild.nativeElement.style.display = 'none';
                this.reorderIndicatorDownViewChild.nativeElement.style.display = 'none';
                this.draggedColumn.draggable = false;
                this.draggedColumn = null;
                this.dropPosition = null;
            }
        };
        TreeTable.prototype.handleRowClick = function (event) {
            var targetNode = event.originalEvent.target.nodeName;
            if (targetNode == 'INPUT' || targetNode == 'BUTTON' || targetNode == 'A' || (dom.DomHandler.hasClass(event.originalEvent.target, 'ui-clickable'))) {
                return;
            }
            if (this.selectionMode) {
                this.preventSelectionSetterPropagation = true;
                var rowNode = event.rowNode;
                var selected = this.isSelected(rowNode.node);
                var metaSelection = this.rowTouched ? false : this.metaKeySelection;
                var dataKeyValue = this.dataKey ? String(utils.ObjectUtils.resolveFieldData(rowNode.node.data, this.dataKey)) : null;
                if (metaSelection) {
                    var metaKey = event.originalEvent.metaKey || event.originalEvent.ctrlKey;
                    if (selected && metaKey) {
                        if (this.isSingleSelectionMode()) {
                            this._selection = null;
                            this.selectionKeys = {};
                            this.selectionChange.emit(null);
                        }
                        else {
                            var selectionIndex_1 = this.findIndexInSelection(rowNode.node);
                            this._selection = this.selection.filter(function (val, i) { return i != selectionIndex_1; });
                            this.selectionChange.emit(this.selection);
                            if (dataKeyValue) {
                                delete this.selectionKeys[dataKeyValue];
                            }
                        }
                        this.onNodeUnselect.emit({ originalEvent: event.originalEvent, node: rowNode.node, type: 'row' });
                    }
                    else {
                        if (this.isSingleSelectionMode()) {
                            this._selection = rowNode.node;
                            this.selectionChange.emit(rowNode.node);
                            if (dataKeyValue) {
                                this.selectionKeys = {};
                                this.selectionKeys[dataKeyValue] = 1;
                            }
                        }
                        else if (this.isMultipleSelectionMode()) {
                            if (metaKey) {
                                this._selection = this.selection || [];
                            }
                            else {
                                this._selection = [];
                                this.selectionKeys = {};
                            }
                            this._selection = __spread(this.selection, [rowNode.node]);
                            this.selectionChange.emit(this.selection);
                            if (dataKeyValue) {
                                this.selectionKeys[dataKeyValue] = 1;
                            }
                        }
                        this.onNodeSelect.emit({ originalEvent: event.originalEvent, node: rowNode.node, type: 'row', index: event.rowIndex });
                    }
                }
                else {
                    if (this.selectionMode === 'single') {
                        if (selected) {
                            this._selection = null;
                            this.selectionKeys = {};
                            this.selectionChange.emit(this.selection);
                            this.onNodeUnselect.emit({ originalEvent: event.originalEvent, node: rowNode.node, type: 'row' });
                        }
                        else {
                            this._selection = rowNode.node;
                            this.selectionChange.emit(this.selection);
                            this.onNodeSelect.emit({ originalEvent: event.originalEvent, node: rowNode.node, type: 'row', index: event.rowIndex });
                            if (dataKeyValue) {
                                this.selectionKeys = {};
                                this.selectionKeys[dataKeyValue] = 1;
                            }
                        }
                    }
                    else if (this.selectionMode === 'multiple') {
                        if (selected) {
                            var selectionIndex_2 = this.findIndexInSelection(rowNode.node);
                            this._selection = this.selection.filter(function (val, i) { return i != selectionIndex_2; });
                            this.selectionChange.emit(this.selection);
                            this.onNodeUnselect.emit({ originalEvent: event.originalEvent, node: rowNode.node, type: 'row' });
                            if (dataKeyValue) {
                                delete this.selectionKeys[dataKeyValue];
                            }
                        }
                        else {
                            this._selection = this.selection ? __spread(this.selection, [rowNode.node]) : [rowNode.node];
                            this.selectionChange.emit(this.selection);
                            this.onNodeSelect.emit({ originalEvent: event.originalEvent, node: rowNode.node, type: 'row', index: event.rowIndex });
                            if (dataKeyValue) {
                                this.selectionKeys[dataKeyValue] = 1;
                            }
                        }
                    }
                }
                this.tableService.onSelectionChange();
            }
            this.rowTouched = false;
        };
        TreeTable.prototype.handleRowTouchEnd = function (event) {
            this.rowTouched = true;
        };
        TreeTable.prototype.handleRowRightClick = function (event) {
            if (this.contextMenu) {
                var node = event.rowNode.node;
                if (this.contextMenuSelectionMode === 'separate') {
                    this.contextMenuSelection = node;
                    this.contextMenuSelectionChange.emit(node);
                    this.onContextMenuSelect.emit({ originalEvent: event.originalEvent, node: node });
                    this.contextMenu.show(event.originalEvent);
                    this.tableService.onContextMenu(node);
                }
                else if (this.contextMenuSelectionMode === 'joint') {
                    this.preventSelectionSetterPropagation = true;
                    var selected = this.isSelected(node);
                    var dataKeyValue = this.dataKey ? String(utils.ObjectUtils.resolveFieldData(node.data, this.dataKey)) : null;
                    if (!selected) {
                        if (this.isSingleSelectionMode()) {
                            this.selection = node;
                            this.selectionChange.emit(node);
                        }
                        else if (this.isMultipleSelectionMode()) {
                            this.selection = [node];
                            this.selectionChange.emit(this.selection);
                        }
                        if (dataKeyValue) {
                            this.selectionKeys[dataKeyValue] = 1;
                        }
                    }
                    this.contextMenu.show(event.originalEvent);
                    this.onContextMenuSelect.emit({ originalEvent: event.originalEvent, node: node });
                }
            }
        };
        TreeTable.prototype.toggleNodeWithCheckbox = function (event) {
            this.selection = this.selection || [];
            this.preventSelectionSetterPropagation = true;
            var node = event.rowNode.node;
            var selected = this.isSelected(node);
            if (selected) {
                this.propagateSelectionDown(node, false);
                if (event.rowNode.parent) {
                    this.propagateSelectionUp(node.parent, false);
                }
                this.selectionChange.emit(this.selection);
                this.onNodeUnselect.emit({ originalEvent: event, node: node });
            }
            else {
                this.propagateSelectionDown(node, true);
                if (event.rowNode.parent) {
                    this.propagateSelectionUp(node.parent, true);
                }
                this.selectionChange.emit(this.selection);
                this.onNodeSelect.emit({ originalEvent: event, node: node });
            }
            this.tableService.onSelectionChange();
        };
        TreeTable.prototype.toggleNodesWithCheckbox = function (event, check) {
            var e_5, _a;
            var data = this.filteredNodes || this.value;
            this._selection = check && data ? data.slice() : [];
            if (check) {
                if (data && data.length) {
                    try {
                        for (var data_1 = __values(data), data_1_1 = data_1.next(); !data_1_1.done; data_1_1 = data_1.next()) {
                            var node = data_1_1.value;
                            this.propagateSelectionDown(node, true);
                        }
                    }
                    catch (e_5_1) { e_5 = { error: e_5_1 }; }
                    finally {
                        try {
                            if (data_1_1 && !data_1_1.done && (_a = data_1.return)) _a.call(data_1);
                        }
                        finally { if (e_5) throw e_5.error; }
                    }
                }
            }
            else {
                this._selection = [];
                this.selectionKeys = {};
            }
            this.preventSelectionSetterPropagation = true;
            this.selectionChange.emit(this._selection);
            this.tableService.onSelectionChange();
            this.onHeaderCheckboxToggle.emit({ originalEvent: event, checked: check });
        };
        TreeTable.prototype.propagateSelectionUp = function (node, select) {
            var e_6, _a;
            if (node.children && node.children.length) {
                var selectedChildCount = 0;
                var childPartialSelected = false;
                var dataKeyValue = this.dataKey ? String(utils.ObjectUtils.resolveFieldData(node.data, this.dataKey)) : null;
                try {
                    for (var _b = __values(node.children), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var child = _c.value;
                        if (this.isSelected(child))
                            selectedChildCount++;
                        else if (child.partialSelected)
                            childPartialSelected = true;
                    }
                }
                catch (e_6_1) { e_6 = { error: e_6_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_6) throw e_6.error; }
                }
                if (select && selectedChildCount == node.children.length) {
                    this._selection = __spread(this.selection || [], [node]);
                    node.partialSelected = false;
                    if (dataKeyValue) {
                        this.selectionKeys[dataKeyValue] = 1;
                    }
                }
                else {
                    if (!select) {
                        var index_1 = this.findIndexInSelection(node);
                        if (index_1 >= 0) {
                            this._selection = this.selection.filter(function (val, i) { return i != index_1; });
                            if (dataKeyValue) {
                                delete this.selectionKeys[dataKeyValue];
                            }
                        }
                    }
                    if (childPartialSelected || selectedChildCount > 0 && selectedChildCount != node.children.length)
                        node.partialSelected = true;
                    else
                        node.partialSelected = false;
                }
            }
            var parent = node.parent;
            if (parent) {
                this.propagateSelectionUp(parent, select);
            }
        };
        TreeTable.prototype.propagateSelectionDown = function (node, select) {
            var e_7, _a;
            var index = this.findIndexInSelection(node);
            var dataKeyValue = this.dataKey ? String(utils.ObjectUtils.resolveFieldData(node.data, this.dataKey)) : null;
            if (select && index == -1) {
                this._selection = __spread(this.selection || [], [node]);
                if (dataKeyValue) {
                    this.selectionKeys[dataKeyValue] = 1;
                }
            }
            else if (!select && index > -1) {
                this._selection = this.selection.filter(function (val, i) { return i != index; });
                if (dataKeyValue) {
                    delete this.selectionKeys[dataKeyValue];
                }
            }
            node.partialSelected = false;
            if (node.children && node.children.length) {
                try {
                    for (var _b = __values(node.children), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var child = _c.value;
                        this.propagateSelectionDown(child, select);
                    }
                }
                catch (e_7_1) { e_7 = { error: e_7_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_7) throw e_7.error; }
                }
            }
        };
        TreeTable.prototype.isSelected = function (node) {
            if (node && this.selection) {
                if (this.dataKey) {
                    return this.selectionKeys[utils.ObjectUtils.resolveFieldData(node.data, this.dataKey)] !== undefined;
                }
                else {
                    if (this.selection instanceof Array)
                        return this.findIndexInSelection(node) > -1;
                    else
                        return this.equals(node, this.selection);
                }
            }
            return false;
        };
        TreeTable.prototype.findIndexInSelection = function (node) {
            var index = -1;
            if (this.selection && this.selection.length) {
                for (var i = 0; i < this.selection.length; i++) {
                    if (this.equals(node, this.selection[i])) {
                        index = i;
                        break;
                    }
                }
            }
            return index;
        };
        TreeTable.prototype.isSingleSelectionMode = function () {
            return this.selectionMode === 'single';
        };
        TreeTable.prototype.isMultipleSelectionMode = function () {
            return this.selectionMode === 'multiple';
        };
        TreeTable.prototype.equals = function (node1, node2) {
            return this.compareSelectionBy === 'equals' ? (node1 === node2) : utils.ObjectUtils.equals(node1.data, node2.data, this.dataKey);
        };
        TreeTable.prototype.filter = function (value, field, matchMode) {
            var _this = this;
            if (this.filterTimeout) {
                clearTimeout(this.filterTimeout);
            }
            if (!this.isFilterBlank(value)) {
                this.filters[field] = { value: value, matchMode: matchMode };
            }
            else if (this.filters[field]) {
                delete this.filters[field];
            }
            this.filterTimeout = setTimeout(function () {
                _this._filter();
                _this.filterTimeout = null;
            }, this.filterDelay);
        };
        TreeTable.prototype.filterGlobal = function (value, matchMode) {
            this.filter(value, 'global', matchMode);
        };
        TreeTable.prototype.isFilterBlank = function (filter) {
            if (filter !== null && filter !== undefined) {
                if ((typeof filter === 'string' && filter.trim().length == 0) || (filter instanceof Array && filter.length == 0))
                    return true;
                else
                    return false;
            }
            return true;
        };
        TreeTable.prototype._filter = function () {
            var e_8, _a;
            if (this.lazy) {
                this.onLazyLoad.emit(this.createLazyLoadMetadata());
            }
            else {
                if (!this.value) {
                    return;
                }
                if (!this.hasFilter()) {
                    this.filteredNodes = null;
                    if (this.paginator) {
                        this.totalRecords = this.value ? this.value.length : 0;
                    }
                }
                else {
                    var globalFilterFieldsArray = void 0;
                    if (this.filters['global']) {
                        if (!this.columns && !this.globalFilterFields)
                            throw new Error('Global filtering requires dynamic columns or globalFilterFields to be defined.');
                        else
                            globalFilterFieldsArray = this.globalFilterFields || this.columns;
                    }
                    this.filteredNodes = [];
                    var isStrictMode = this.filterMode === 'strict';
                    var isValueChanged = false;
                    try {
                        for (var _b = __values(this.value), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var node = _c.value;
                            var copyNode = __assign({}, node);
                            var localMatch = true;
                            var globalMatch = false;
                            var paramsWithoutNode = void 0;
                            for (var prop in this.filters) {
                                if (this.filters.hasOwnProperty(prop) && prop !== 'global') {
                                    var filterMeta = this.filters[prop];
                                    var filterField = prop;
                                    var filterValue = filterMeta.value;
                                    var filterMatchMode = filterMeta.matchMode || 'startsWith';
                                    var filterConstraint = utils.FilterUtils[filterMatchMode];
                                    paramsWithoutNode = { filterField: filterField, filterValue: filterValue, filterConstraint: filterConstraint, isStrictMode: isStrictMode };
                                    if ((isStrictMode && !(this.findFilteredNodes(copyNode, paramsWithoutNode) || this.isFilterMatched(copyNode, paramsWithoutNode))) ||
                                        (!isStrictMode && !(this.isFilterMatched(copyNode, paramsWithoutNode) || this.findFilteredNodes(copyNode, paramsWithoutNode)))) {
                                        localMatch = false;
                                    }
                                    if (!localMatch) {
                                        break;
                                    }
                                }
                            }
                            if (this.filters['global'] && !globalMatch && globalFilterFieldsArray) {
                                for (var j = 0; j < globalFilterFieldsArray.length; j++) {
                                    var copyNodeForGlobal = __assign({}, copyNode);
                                    var filterField = globalFilterFieldsArray[j].field || globalFilterFieldsArray[j];
                                    var filterValue = this.filters['global'].value;
                                    var filterConstraint = utils.FilterUtils[this.filters['global'].matchMode];
                                    paramsWithoutNode = { filterField: filterField, filterValue: filterValue, filterConstraint: filterConstraint, isStrictMode: isStrictMode };
                                    if ((isStrictMode && (this.findFilteredNodes(copyNodeForGlobal, paramsWithoutNode) || this.isFilterMatched(copyNodeForGlobal, paramsWithoutNode))) ||
                                        (!isStrictMode && (this.isFilterMatched(copyNodeForGlobal, paramsWithoutNode) || this.findFilteredNodes(copyNodeForGlobal, paramsWithoutNode)))) {
                                        globalMatch = true;
                                        copyNode = copyNodeForGlobal;
                                    }
                                }
                            }
                            var matches = localMatch;
                            if (this.filters['global']) {
                                matches = localMatch && globalMatch;
                            }
                            if (matches) {
                                this.filteredNodes.push(copyNode);
                            }
                            isValueChanged = isValueChanged || !localMatch || globalMatch || (localMatch && this.filteredNodes.length > 0) || (!globalMatch && this.filteredNodes.length === 0);
                        }
                    }
                    catch (e_8_1) { e_8 = { error: e_8_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_8) throw e_8.error; }
                    }
                    if (!isValueChanged) {
                        this.filteredNodes = null;
                    }
                    if (this.paginator) {
                        this.totalRecords = this.filteredNodes ? this.filteredNodes.length : this.value ? this.value.length : 0;
                    }
                }
            }
            this.first = 0;
            var filteredValue = this.filteredNodes || this.value;
            this.onFilter.emit({
                filters: this.filters,
                filteredValue: filteredValue
            });
            this.tableService.onUIUpdate(filteredValue);
            this.updateSerializedValue();
            if (this.scrollable) {
                this.resetScrollTop();
            }
        };
        TreeTable.prototype.findFilteredNodes = function (node, paramsWithoutNode) {
            var e_9, _a;
            if (node) {
                var matched = false;
                if (node.children) {
                    var childNodes = __spread(node.children);
                    node.children = [];
                    try {
                        for (var childNodes_1 = __values(childNodes), childNodes_1_1 = childNodes_1.next(); !childNodes_1_1.done; childNodes_1_1 = childNodes_1.next()) {
                            var childNode = childNodes_1_1.value;
                            var copyChildNode = __assign({}, childNode);
                            if (this.isFilterMatched(copyChildNode, paramsWithoutNode)) {
                                matched = true;
                                node.children.push(copyChildNode);
                            }
                        }
                    }
                    catch (e_9_1) { e_9 = { error: e_9_1 }; }
                    finally {
                        try {
                            if (childNodes_1_1 && !childNodes_1_1.done && (_a = childNodes_1.return)) _a.call(childNodes_1);
                        }
                        finally { if (e_9) throw e_9.error; }
                    }
                }
                if (matched) {
                    return true;
                }
            }
        };
        TreeTable.prototype.isFilterMatched = function (node, _a) {
            var filterField = _a.filterField, filterValue = _a.filterValue, filterConstraint = _a.filterConstraint, isStrictMode = _a.isStrictMode;
            var matched = false;
            var dataFieldValue = utils.ObjectUtils.resolveFieldData(node.data, filterField);
            if (filterConstraint(dataFieldValue, filterValue, this.filterLocale)) {
                matched = true;
            }
            if (!matched || (isStrictMode && !this.isNodeLeaf(node))) {
                matched = this.findFilteredNodes(node, { filterField: filterField, filterValue: filterValue, filterConstraint: filterConstraint, isStrictMode: isStrictMode }) || matched;
            }
            return matched;
        };
        TreeTable.prototype.isNodeLeaf = function (node) {
            return node.leaf === false ? false : !(node.children && node.children.length);
        };
        TreeTable.prototype.hasFilter = function () {
            var empty = true;
            for (var prop in this.filters) {
                if (this.filters.hasOwnProperty(prop)) {
                    empty = false;
                    break;
                }
            }
            return !empty;
        };
        TreeTable.prototype.reset = function () {
            this._sortField = null;
            this._sortOrder = 1;
            this._multiSortMeta = null;
            this.tableService.onSort(null);
            this.filteredNodes = null;
            this.filters = {};
            this.first = 0;
            if (this.lazy) {
                this.onLazyLoad.emit(this.createLazyLoadMetadata());
            }
            else {
                this.totalRecords = (this._value ? this._value.length : 0);
            }
        };
        TreeTable.prototype.updateEditingCell = function (cell) {
            this.editingCell = cell;
            this.bindDocumentEditListener();
        };
        TreeTable.prototype.isEditingCellValid = function () {
            return (this.editingCell && dom.DomHandler.find(this.editingCell, '.ng-invalid.ng-dirty').length === 0);
        };
        TreeTable.prototype.bindDocumentEditListener = function () {
            var _this = this;
            if (!this.documentEditListener) {
                this.documentEditListener = function (event) {
                    if (_this.editingCell && !_this.editingCellClick && _this.isEditingCellValid()) {
                        dom.DomHandler.removeClass(_this.editingCell, 'ui-editing-cell');
                        _this.editingCell = null;
                        _this.unbindDocumentEditListener();
                    }
                    _this.editingCellClick = false;
                };
                document.addEventListener('click', this.documentEditListener);
            }
        };
        TreeTable.prototype.unbindDocumentEditListener = function () {
            if (this.documentEditListener) {
                document.removeEventListener('click', this.documentEditListener);
                this.documentEditListener = null;
            }
        };
        TreeTable.prototype.ngOnDestroy = function () {
            this.unbindDocumentEditListener();
            this.editingCell = null;
            this.initialized = null;
        };
        TreeTable.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: core.NgZone },
            { type: TreeTableService }
        ]; };
        __decorate([
            core.Input()
        ], TreeTable.prototype, "columns", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "style", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "styleClass", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "tableStyle", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "tableStyleClass", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "autoLayout", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "lazy", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "lazyLoadOnInit", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "paginator", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "rows", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "first", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "pageLinks", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "rowsPerPageOptions", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "alwaysShowPaginator", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "paginatorPosition", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "paginatorDropdownAppendTo", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "currentPageReportTemplate", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "showCurrentPageReport", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "defaultSortOrder", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "sortMode", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "resetPageOnSort", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "customSort", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "selectionMode", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "selectionChange", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "contextMenuSelection", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "contextMenuSelectionChange", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "contextMenuSelectionMode", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "dataKey", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "metaKeySelection", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "compareSelectionBy", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "rowHover", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "loading", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "loadingIcon", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "showLoader", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "scrollable", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "scrollHeight", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "virtualScroll", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "virtualScrollDelay", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "virtualRowHeight", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "minBufferPx", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "maxBufferPx", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "frozenWidth", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "frozenColumns", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "resizableColumns", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "columnResizeMode", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "reorderableColumns", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "contextMenu", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "rowTrackBy", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "filters", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "globalFilterFields", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "filterDelay", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "filterMode", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "filterLocale", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onFilter", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onNodeExpand", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onNodeCollapse", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onPage", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onSort", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onLazyLoad", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "sortFunction", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onColResize", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onColReorder", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onNodeSelect", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onNodeUnselect", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onContextMenuSelect", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onHeaderCheckboxToggle", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onEditInit", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onEditComplete", void 0);
        __decorate([
            core.Output()
        ], TreeTable.prototype, "onEditCancel", void 0);
        __decorate([
            core.ViewChild('container')
        ], TreeTable.prototype, "containerViewChild", void 0);
        __decorate([
            core.ViewChild('resizeHelper')
        ], TreeTable.prototype, "resizeHelperViewChild", void 0);
        __decorate([
            core.ViewChild('reorderIndicatorUp')
        ], TreeTable.prototype, "reorderIndicatorUpViewChild", void 0);
        __decorate([
            core.ViewChild('reorderIndicatorDown')
        ], TreeTable.prototype, "reorderIndicatorDownViewChild", void 0);
        __decorate([
            core.ViewChild('table')
        ], TreeTable.prototype, "tableViewChild", void 0);
        __decorate([
            core.ViewChild('scrollableView')
        ], TreeTable.prototype, "scrollableViewChild", void 0);
        __decorate([
            core.ViewChild('scrollableFrozenView')
        ], TreeTable.prototype, "scrollableFrozenViewChild", void 0);
        __decorate([
            core.ContentChildren(api.PrimeTemplate)
        ], TreeTable.prototype, "templates", void 0);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "value", null);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "totalRecords", null);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "sortField", null);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "sortOrder", null);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "multiSortMeta", null);
        __decorate([
            core.Input()
        ], TreeTable.prototype, "selection", null);
        TreeTable = __decorate([
            core.Component({
                selector: 'p-treeTable',
                template: "\n        <div #container [ngStyle]=\"style\" [class]=\"styleClass\"\n                [ngClass]=\"{'ui-treetable ui-widget': true, 'ui-treetable-auto-layout': autoLayout, 'ui-treetable-hoverable-rows': (rowHover||(selectionMode === 'single' || selectionMode === 'multiple')),\n                'ui-treetable-resizable': resizableColumns, 'ui-treetable-resizable-fit': (resizableColumns && columnResizeMode === 'fit'), 'ui-treetable-flex-scrollable': (scrollable && scrollHeight === 'flex')}\">\n            <div class=\"ui-treetable-loading ui-widget-overlay\" *ngIf=\"loading && showLoader\"></div>\n            <div class=\"ui-treetable-loading-content\" *ngIf=\"loading && showLoader\">\n                <i [class]=\"'ui-treetable-loading-icon pi-spin ' + loadingIcon\"></i>\n            </div>\n            <div *ngIf=\"captionTemplate\" class=\"ui-treetable-caption ui-widget-header\">\n                <ng-container *ngTemplateOutlet=\"captionTemplate\"></ng-container>\n            </div>\n            <p-paginator [rows]=\"rows\" [first]=\"first\" [totalRecords]=\"totalRecords\" [pageLinkSize]=\"pageLinks\" styleClass=\"ui-paginator-top\" [alwaysShow]=\"alwaysShowPaginator\"\n                (onPageChange)=\"onPageChange($event)\" [rowsPerPageOptions]=\"rowsPerPageOptions\" *ngIf=\"paginator && (paginatorPosition === 'top' || paginatorPosition =='both')\"\n                [templateLeft]=\"paginatorLeftTemplate\" [templateRight]=\"paginatorRightTemplate\" [dropdownAppendTo]=\"paginatorDropdownAppendTo\"\n                [currentPageReportTemplate]=\"currentPageReportTemplate\" [showCurrentPageReport]=\"showCurrentPageReport\"></p-paginator>\n\n            <div class=\"ui-treetable-wrapper\" *ngIf=\"!scrollable\">\n                <table #table class=\"ui-treetable-table\" [ngClass]=\"tableStyleClass\" [ngStyle]=\"tableStyle\">\n                    <ng-container *ngTemplateOutlet=\"colGroupTemplate; context {$implicit: columns}\"></ng-container>\n                    <thead class=\"ui-treetable-thead\">\n                        <ng-container *ngTemplateOutlet=\"headerTemplate; context: {$implicit: columns}\"></ng-container>\n                    </thead>\n                    <tfoot class=\"ui-treetable-tfoot\">\n                        <ng-container *ngTemplateOutlet=\"footerTemplate; context {$implicit: columns}\"></ng-container>\n                    </tfoot>\n                    <tbody class=\"ui-treetable-tbody\" [pTreeTableBody]=\"columns\" [pTreeTableBodyTemplate]=\"bodyTemplate\"></tbody>\n                </table>\n            </div>\n\n            <div class=\"ui-treetable-scrollable-wrapper\" *ngIf=\"scrollable\">\n               <div class=\"ui-treetable-scrollable-view ui-treetable-frozen-view\" *ngIf=\"frozenColumns||frozenBodyTemplate\" #scrollableFrozenView [ttScrollableView]=\"frozenColumns\" [frozen]=\"true\" [ngStyle]=\"{width: frozenWidth}\" [scrollHeight]=\"scrollHeight\"></div>\n               <div class=\"ui-treetable-scrollable-view\" #scrollableView [ttScrollableView]=\"columns\" [frozen]=\"false\" [scrollHeight]=\"scrollHeight\" [ngStyle]=\"{left: frozenWidth, width: 'calc(100% - '+frozenWidth+')'}\"></div>\n            </div>\n\n            <p-paginator [rows]=\"rows\" [first]=\"first\" [totalRecords]=\"totalRecords\" [pageLinkSize]=\"pageLinks\" styleClass=\"ui-paginator-bottom\" [alwaysShow]=\"alwaysShowPaginator\"\n                (onPageChange)=\"onPageChange($event)\" [rowsPerPageOptions]=\"rowsPerPageOptions\" *ngIf=\"paginator && (paginatorPosition === 'bottom' || paginatorPosition =='both')\"\n                [templateLeft]=\"paginatorLeftTemplate\" [templateRight]=\"paginatorRightTemplate\" [dropdownAppendTo]=\"paginatorDropdownAppendTo\"\n                [currentPageReportTemplate]=\"currentPageReportTemplate\" [showCurrentPageReport]=\"showCurrentPageReport\"></p-paginator>\n            <div *ngIf=\"summaryTemplate\" class=\"ui-treetable-summary ui-widget-header\">\n                <ng-container *ngTemplateOutlet=\"summaryTemplate\"></ng-container>\n            </div>\n\n            <div #resizeHelper class=\"ui-column-resizer-helper ui-state-highlight\" style=\"display:none\" *ngIf=\"resizableColumns\"></div>\n\n            <span #reorderIndicatorUp class=\"pi pi-arrow-down ui-table-reorder-indicator-up\" *ngIf=\"reorderableColumns\"></span>\n            <span #reorderIndicatorDown class=\"pi pi-arrow-up ui-table-reorder-indicator-down\" *ngIf=\"reorderableColumns\"></span>\n        </div>\n    ",
                providers: [TreeTableService],
                changeDetection: core.ChangeDetectionStrategy.Default
            })
        ], TreeTable);
        return TreeTable;
    }());
    var TTBody = /** @class */ (function () {
        function TTBody(tt) {
            this.tt = tt;
        }
        TTBody.ctorParameters = function () { return [
            { type: TreeTable }
        ]; };
        __decorate([
            core.Input("pTreeTableBody")
        ], TTBody.prototype, "columns", void 0);
        __decorate([
            core.Input("pTreeTableBodyTemplate")
        ], TTBody.prototype, "template", void 0);
        __decorate([
            core.Input()
        ], TTBody.prototype, "frozen", void 0);
        TTBody = __decorate([
            core.Component({
                selector: '[pTreeTableBody]',
                template: "\n        <ng-container *ngIf=\"!tt.virtualScroll\">\n            <ng-template ngFor let-serializedNode let-rowIndex=\"index\" [ngForOf]=\"tt.serializedValue\" [ngForTrackBy]=\"tt.rowTrackBy\">\n                <ng-container *ngIf=\"serializedNode.visible\">\n                    <ng-container *ngTemplateOutlet=\"template; context: {$implicit: serializedNode, node: serializedNode.node, rowData: serializedNode.node.data, columns: columns}\"></ng-container>\n                </ng-container>\n            </ng-template>\n        </ng-container>\n        <ng-container *ngIf=\"tt.virtualScroll\">\n            <ng-template cdkVirtualFor let-serializedNode let-rowIndex=\"index\" [cdkVirtualForOf]=\"tt.serializedValue\" [cdkVirtualForTrackBy]=\"tt.rowTrackBy\">\n                <ng-container *ngIf=\"serializedNode.visible\">\n                    <ng-container *ngTemplateOutlet=\"template; context: {$implicit: serializedNode, node: serializedNode.node, rowData: serializedNode.node.data, columns: columns}\"></ng-container>\n                </ng-container>\n            </ng-template>\n        </ng-container>\n        <ng-container *ngIf=\"tt.isEmpty()\">\n            <ng-container *ngTemplateOutlet=\"tt.emptyMessageTemplate; context: {$implicit: columns}\"></ng-container>\n        </ng-container>\n    "
            })
        ], TTBody);
        return TTBody;
    }());
    var TTScrollableView = /** @class */ (function () {
        function TTScrollableView(tt, el, zone) {
            var _this = this;
            this.tt = tt;
            this.el = el;
            this.zone = zone;
            this.subscription = this.tt.tableService.uiUpdateSource$.subscribe(function () {
                _this.zone.runOutsideAngular(function () {
                    setTimeout(function () {
                        _this.alignScrollBar();
                        _this.initialized = true;
                    }, 50);
                });
            });
            this.initialized = false;
        }
        Object.defineProperty(TTScrollableView.prototype, "scrollHeight", {
            get: function () {
                return this._scrollHeight;
            },
            set: function (val) {
                this._scrollHeight = val;
                if (val != null && (val.includes('%') || val.includes('calc'))) {
                    console.log('Percentage scroll height calculation is removed in favor of the more performant CSS based flex mode, use scrollHeight="flex" instead.');
                }
            },
            enumerable: true,
            configurable: true
        });
        TTScrollableView.prototype.ngAfterViewChecked = function () {
            if (!this.initialized && this.el.nativeElement.offsetParent) {
                this.alignScrollBar();
                this.initialized = true;
            }
        };
        TTScrollableView.prototype.ngAfterViewInit = function () {
            if (!this.frozen) {
                if (this.tt.frozenColumns || this.tt.frozenBodyTemplate) {
                    dom.DomHandler.addClass(this.el.nativeElement, 'ui-treetable-unfrozen-view');
                }
                var frozenView = this.el.nativeElement.previousElementSibling;
                if (frozenView) {
                    if (this.tt.virtualScroll)
                        this.frozenSiblingBody = dom.DomHandler.findSingle(frozenView, '.ui-treetable-virtual-scrollable-body');
                    else
                        this.frozenSiblingBody = dom.DomHandler.findSingle(frozenView, '.ui-treetable-scrollable-body');
                }
            }
            else {
                if (this.scrollableAlignerViewChild && this.scrollableAlignerViewChild.nativeElement) {
                    this.scrollableAlignerViewChild.nativeElement.style.height = dom.DomHandler.calculateScrollbarHeight() + 'px';
                }
            }
            this.bindEvents();
            this.alignScrollBar();
        };
        TTScrollableView.prototype.bindEvents = function () {
            var _this = this;
            this.zone.runOutsideAngular(function () {
                var scrollBarWidth = dom.DomHandler.calculateScrollbarWidth();
                if (_this.scrollHeaderViewChild && _this.scrollHeaderViewChild.nativeElement) {
                    _this.headerScrollListener = _this.onHeaderScroll.bind(_this);
                    _this.scrollHeaderBoxViewChild.nativeElement.addEventListener('scroll', _this.headerScrollListener);
                }
                if (_this.scrollFooterViewChild && _this.scrollFooterViewChild.nativeElement) {
                    _this.footerScrollListener = _this.onFooterScroll.bind(_this);
                    _this.scrollFooterViewChild.nativeElement.addEventListener('scroll', _this.footerScrollListener);
                }
                if (!_this.frozen) {
                    _this.bodyScrollListener = _this.onBodyScroll.bind(_this);
                    if (_this.tt.virtualScroll)
                        _this.virtualScrollBody.getElementRef().nativeElement.addEventListener('scroll', _this.bodyScrollListener);
                    else
                        _this.scrollBodyViewChild.nativeElement.addEventListener('scroll', _this.bodyScrollListener);
                }
            });
        };
        TTScrollableView.prototype.unbindEvents = function () {
            if (this.scrollHeaderViewChild && this.scrollHeaderViewChild.nativeElement) {
                this.scrollHeaderBoxViewChild.nativeElement.removeEventListener('scroll', this.headerScrollListener);
            }
            if (this.scrollFooterViewChild && this.scrollFooterViewChild.nativeElement) {
                this.scrollFooterViewChild.nativeElement.removeEventListener('scroll', this.footerScrollListener);
            }
            if (this.scrollBodyViewChild && this.scrollBodyViewChild.nativeElement) {
                this.scrollBodyViewChild.nativeElement.removeEventListener('scroll', this.bodyScrollListener);
            }
            if (this.virtualScrollBody && this.virtualScrollBody.getElementRef()) {
                this.virtualScrollBody.getElementRef().nativeElement.removeEventListener('scroll', this.bodyScrollListener);
            }
        };
        TTScrollableView.prototype.onHeaderScroll = function () {
            var scrollLeft = this.scrollHeaderViewChild.nativeElement.scrollLeft;
            this.scrollBodyViewChild.nativeElement.scrollLeft = scrollLeft;
            if (this.scrollFooterViewChild && this.scrollFooterViewChild.nativeElement) {
                this.scrollFooterViewChild.nativeElement.scrollLeft = scrollLeft;
            }
            this.preventBodyScrollPropagation = true;
        };
        TTScrollableView.prototype.onFooterScroll = function () {
            var scrollLeft = this.scrollFooterViewChild.nativeElement.scrollLeft;
            this.scrollBodyViewChild.nativeElement.scrollLeft = scrollLeft;
            if (this.scrollHeaderViewChild && this.scrollHeaderViewChild.nativeElement) {
                this.scrollHeaderViewChild.nativeElement.scrollLeft = scrollLeft;
            }
            this.preventBodyScrollPropagation = true;
        };
        TTScrollableView.prototype.onBodyScroll = function (event) {
            if (this.preventBodyScrollPropagation) {
                this.preventBodyScrollPropagation = false;
                return;
            }
            if (this.scrollHeaderViewChild && this.scrollHeaderViewChild.nativeElement) {
                this.scrollHeaderBoxViewChild.nativeElement.style.marginLeft = -1 * event.target.scrollLeft + 'px';
            }
            if (this.scrollFooterViewChild && this.scrollFooterViewChild.nativeElement) {
                this.scrollFooterBoxViewChild.nativeElement.style.marginLeft = -1 * event.target.scrollLeft + 'px';
            }
            if (this.frozenSiblingBody) {
                this.frozenSiblingBody.scrollTop = event.target.scrollTop;
            }
        };
        TTScrollableView.prototype.scrollToVirtualIndex = function (index) {
            if (this.virtualScrollBody) {
                this.virtualScrollBody.scrollToIndex(index);
            }
        };
        TTScrollableView.prototype.scrollTo = function (options) {
            if (this.virtualScrollBody) {
                this.virtualScrollBody.scrollTo(options);
            }
            else {
                if (this.scrollBodyViewChild.nativeElement.scrollTo) {
                    this.scrollBodyViewChild.nativeElement.scrollTo(options);
                }
                else {
                    this.scrollBodyViewChild.nativeElement.scrollLeft = options.left;
                    this.scrollBodyViewChild.nativeElement.scrollTop = options.top;
                }
            }
        };
        TTScrollableView.prototype.hasVerticalOverflow = function () {
            if (this.tt.virtualScroll)
                return (this.virtualScrollBody.getDataLength() * this.tt.virtualRowHeight) > this.virtualScrollBody.getViewportSize();
            else
                return dom.DomHandler.getOuterHeight(this.scrollTableViewChild.nativeElement) > dom.DomHandler.getOuterHeight(this.scrollBodyViewChild.nativeElement);
        };
        TTScrollableView.prototype.alignScrollBar = function () {
            if (!this.frozen) {
                var scrollBarWidth = this.hasVerticalOverflow() ? dom.DomHandler.calculateScrollbarWidth() : 0;
                this.scrollHeaderBoxViewChild.nativeElement.style.marginRight = scrollBarWidth + 'px';
                if (this.scrollFooterBoxViewChild && this.scrollFooterBoxViewChild.nativeElement) {
                    this.scrollFooterBoxViewChild.nativeElement.style.marginRight = scrollBarWidth + 'px';
                }
            }
            this.initialized = false;
        };
        TTScrollableView.prototype.ngOnDestroy = function () {
            this.unbindEvents();
            this.frozenSiblingBody = null;
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
            this.initialized = false;
        };
        TTScrollableView.ctorParameters = function () { return [
            { type: TreeTable },
            { type: core.ElementRef },
            { type: core.NgZone }
        ]; };
        __decorate([
            core.Input("ttScrollableView")
        ], TTScrollableView.prototype, "columns", void 0);
        __decorate([
            core.Input()
        ], TTScrollableView.prototype, "frozen", void 0);
        __decorate([
            core.ViewChild('scrollHeader')
        ], TTScrollableView.prototype, "scrollHeaderViewChild", void 0);
        __decorate([
            core.ViewChild('scrollHeaderBox')
        ], TTScrollableView.prototype, "scrollHeaderBoxViewChild", void 0);
        __decorate([
            core.ViewChild('scrollBody')
        ], TTScrollableView.prototype, "scrollBodyViewChild", void 0);
        __decorate([
            core.ViewChild('scrollTable')
        ], TTScrollableView.prototype, "scrollTableViewChild", void 0);
        __decorate([
            core.ViewChild('loadingTable')
        ], TTScrollableView.prototype, "scrollLoadingTableViewChild", void 0);
        __decorate([
            core.ViewChild('scrollFooter')
        ], TTScrollableView.prototype, "scrollFooterViewChild", void 0);
        __decorate([
            core.ViewChild('scrollFooterBox')
        ], TTScrollableView.prototype, "scrollFooterBoxViewChild", void 0);
        __decorate([
            core.ViewChild('scrollableAligner')
        ], TTScrollableView.prototype, "scrollableAlignerViewChild", void 0);
        __decorate([
            core.ViewChild(scrolling.CdkVirtualScrollViewport)
        ], TTScrollableView.prototype, "virtualScrollBody", void 0);
        __decorate([
            core.Input()
        ], TTScrollableView.prototype, "scrollHeight", null);
        TTScrollableView = __decorate([
            core.Component({
                selector: '[ttScrollableView]',
                template: "\n        <div #scrollHeader class=\"ui-treetable-scrollable-header ui-widget-header\">\n            <div #scrollHeaderBox class=\"ui-treetable-scrollable-header-box\">\n                <table class=\"ui-treetable-scrollable-header-table\" [ngClass]=\"tt.tableStyleClass\" [ngStyle]=\"tt.tableStyle\">\n                    <ng-container *ngTemplateOutlet=\"frozen ? tt.frozenColGroupTemplate||tt.colGroupTemplate : tt.colGroupTemplate; context {$implicit: columns}\"></ng-container>\n                    <thead class=\"ui-treetable-thead\">\n                        <ng-container *ngTemplateOutlet=\"frozen ? tt.frozenHeaderTemplate||tt.headerTemplate : tt.headerTemplate; context {$implicit: columns}\"></ng-container>\n                    </thead>\n                </table>\n            </div>\n        </div>\n        <ng-container *ngIf=\"!tt.virtualScroll; else virtualScrollTemplate\">\n            <div #scrollBody class=\"ui-treetable-scrollable-body\" [ngStyle]=\"{'max-height': tt.scrollHeight !== 'flex' ? scrollHeight : undefined}\">\n                <table #scrollTable [class]=\"tt.tableStyleClass\" [ngStyle]=\"tt.tableStyle\">\n                    <ng-container *ngTemplateOutlet=\"frozen ? tt.frozenColGroupTemplate||tt.colGroupTemplate : tt.colGroupTemplate; context {$implicit: columns}\"></ng-container>\n                    <tbody class=\"ui-treetable-tbody\" [pTreeTableBody]=\"columns\" [pTreeTableBodyTemplate]=\"frozen ? tt.frozenBodyTemplate||tt.bodyTemplate : tt.bodyTemplate\" [frozen]=\"frozen\"></tbody>\n                </table>\n                <div #scrollableAligner style=\"background-color:transparent\" *ngIf=\"frozen\"></div>\n            </div>\n        </ng-container>\n        <ng-template #virtualScrollTemplate>\n            <cdk-virtual-scroll-viewport [itemSize]=\"tt.virtualRowHeight\" [style.height]=\"tt.scrollHeight !== 'flex' ? scrollHeight : undefined\" \n                    [minBufferPx]=\"tt.minBufferPx\" [maxBufferPx]=\"tt.maxBufferPx\" class=\"ui-treetable-virtual-scrollable-body\">\n                <table #scrollTable [class]=\"tt.tableStyleClass\" [ngStyle]=\"tt.tableStyle\">\n                    <ng-container *ngTemplateOutlet=\"frozen ? tt.frozenColGroupTemplate||tt.colGroupTemplate : tt.colGroupTemplate; context {$implicit: columns}\"></ng-container>\n                    <tbody class=\"ui-treetable-tbody\" [pTreeTableBody]=\"columns\" [pTreeTableBodyTemplate]=\"frozen ? tt.frozenBodyTemplate||tt.bodyTemplate : tt.bodyTemplate\" [frozen]=\"frozen\"></tbody>\n                </table>\n                <div #scrollableAligner style=\"background-color:transparent\" *ngIf=\"frozen\"></div>\n            </cdk-virtual-scroll-viewport>\n        </ng-template>\n        <div #scrollFooter *ngIf=\"tt.footerTemplate\" class=\"ui-treetable-scrollable-footer ui-widget-header\">\n            <div #scrollFooterBox class=\"ui-treetable-scrollable-footer-box\">\n                <table class=\"ui-treetable-scrollable-footer-table\" [ngClass]=\"tt.tableStyleClass\" [ngStyle]=\"tt.tableStyle\">\n                    <ng-container *ngTemplateOutlet=\"frozen ? tt.frozenColGroupTemplate||tt.colGroupTemplate : tt.colGroupTemplate; context {$implicit: columns}\"></ng-container>\n                    <tfoot class=\"ui-treetable-tfoot\">\n                        <ng-container *ngTemplateOutlet=\"frozen ? tt.frozenFooterTemplate||tt.footerTemplate : tt.footerTemplate; context {$implicit: columns}\"></ng-container>\n                    </tfoot>\n                </table>\n            </div>\n        </div>\n    "
            })
        ], TTScrollableView);
        return TTScrollableView;
    }());
    var TTSortableColumn = /** @class */ (function () {
        function TTSortableColumn(tt) {
            var _this = this;
            this.tt = tt;
            if (this.isEnabled()) {
                this.subscription = this.tt.tableService.sortSource$.subscribe(function (sortMeta) {
                    _this.updateSortState();
                });
            }
        }
        TTSortableColumn.prototype.ngOnInit = function () {
            if (this.isEnabled()) {
                this.updateSortState();
            }
        };
        TTSortableColumn.prototype.updateSortState = function () {
            this.sorted = this.tt.isSorted(this.field);
        };
        TTSortableColumn.prototype.onClick = function (event) {
            if (this.isEnabled()) {
                this.updateSortState();
                this.tt.sort({
                    originalEvent: event,
                    field: this.field
                });
                dom.DomHandler.clearSelection();
            }
        };
        TTSortableColumn.prototype.onEnterKey = function (event) {
            this.onClick(event);
        };
        TTSortableColumn.prototype.isEnabled = function () {
            return this.ttSortableColumnDisabled !== true;
        };
        TTSortableColumn.prototype.ngOnDestroy = function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        TTSortableColumn.ctorParameters = function () { return [
            { type: TreeTable }
        ]; };
        __decorate([
            core.Input("ttSortableColumn")
        ], TTSortableColumn.prototype, "field", void 0);
        __decorate([
            core.Input()
        ], TTSortableColumn.prototype, "ttSortableColumnDisabled", void 0);
        __decorate([
            core.HostListener('click', ['$event'])
        ], TTSortableColumn.prototype, "onClick", null);
        __decorate([
            core.HostListener('keydown.enter', ['$event'])
        ], TTSortableColumn.prototype, "onEnterKey", null);
        TTSortableColumn = __decorate([
            core.Directive({
                selector: '[ttSortableColumn]',
                host: {
                    '[class.ui-sortable-column]': 'isEnabled()',
                    '[class.ui-state-highlight]': 'sorted',
                    '[attr.tabindex]': 'isEnabled() ? "0" : null'
                }
            })
        ], TTSortableColumn);
        return TTSortableColumn;
    }());
    var TTSortIcon = /** @class */ (function () {
        function TTSortIcon(tt) {
            var _this = this;
            this.tt = tt;
            this.subscription = this.tt.tableService.sortSource$.subscribe(function (sortMeta) {
                _this.updateSortState();
            });
        }
        TTSortIcon.prototype.ngOnInit = function () {
            this.updateSortState();
        };
        TTSortIcon.prototype.onClick = function (event) {
            event.preventDefault();
        };
        TTSortIcon.prototype.updateSortState = function () {
            if (this.tt.sortMode === 'single') {
                this.sortOrder = this.tt.isSorted(this.field) ? this.tt.sortOrder : 0;
            }
            else if (this.tt.sortMode === 'multiple') {
                var sortMeta = this.tt.getSortMeta(this.field);
                this.sortOrder = sortMeta ? sortMeta.order : 0;
            }
        };
        TTSortIcon.prototype.ngOnDestroy = function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        TTSortIcon.ctorParameters = function () { return [
            { type: TreeTable }
        ]; };
        __decorate([
            core.Input()
        ], TTSortIcon.prototype, "field", void 0);
        __decorate([
            core.Input()
        ], TTSortIcon.prototype, "ariaLabelDesc", void 0);
        __decorate([
            core.Input()
        ], TTSortIcon.prototype, "ariaLabelAsc", void 0);
        TTSortIcon = __decorate([
            core.Component({
                selector: 'p-treeTableSortIcon',
                template: "\n        <i class=\"ui-sortable-column-icon pi pi-fw\" [ngClass]=\"{'pi-sort-amount-up-alt': sortOrder === 1, 'pi-sort-amount-down': sortOrder === -1, 'pi-sort-alt': sortOrder === 0}\"></i>\n    "
            })
        ], TTSortIcon);
        return TTSortIcon;
    }());
    var TTResizableColumn = /** @class */ (function () {
        function TTResizableColumn(tt, el, zone) {
            this.tt = tt;
            this.el = el;
            this.zone = zone;
        }
        TTResizableColumn.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (this.isEnabled()) {
                dom.DomHandler.addClass(this.el.nativeElement, 'ui-resizable-column');
                this.resizer = document.createElement('span');
                this.resizer.className = 'ui-column-resizer ui-clickable';
                this.el.nativeElement.appendChild(this.resizer);
                this.zone.runOutsideAngular(function () {
                    _this.resizerMouseDownListener = _this.onMouseDown.bind(_this);
                    _this.resizer.addEventListener('mousedown', _this.resizerMouseDownListener);
                });
            }
        };
        TTResizableColumn.prototype.bindDocumentEvents = function () {
            var _this = this;
            this.zone.runOutsideAngular(function () {
                _this.documentMouseMoveListener = _this.onDocumentMouseMove.bind(_this);
                document.addEventListener('mousemove', _this.documentMouseMoveListener);
                _this.documentMouseUpListener = _this.onDocumentMouseUp.bind(_this);
                document.addEventListener('mouseup', _this.documentMouseUpListener);
            });
        };
        TTResizableColumn.prototype.unbindDocumentEvents = function () {
            if (this.documentMouseMoveListener) {
                document.removeEventListener('mousemove', this.documentMouseMoveListener);
                this.documentMouseMoveListener = null;
            }
            if (this.documentMouseUpListener) {
                document.removeEventListener('mouseup', this.documentMouseUpListener);
                this.documentMouseUpListener = null;
            }
        };
        TTResizableColumn.prototype.onMouseDown = function (event) {
            this.tt.onColumnResizeBegin(event);
            this.bindDocumentEvents();
        };
        TTResizableColumn.prototype.onDocumentMouseMove = function (event) {
            this.tt.onColumnResize(event);
        };
        TTResizableColumn.prototype.onDocumentMouseUp = function (event) {
            this.tt.onColumnResizeEnd(event, this.el.nativeElement);
            this.unbindDocumentEvents();
        };
        TTResizableColumn.prototype.isEnabled = function () {
            return this.ttResizableColumnDisabled !== true;
        };
        TTResizableColumn.prototype.ngOnDestroy = function () {
            if (this.resizerMouseDownListener) {
                this.resizer.removeEventListener('mousedown', this.resizerMouseDownListener);
            }
            this.unbindDocumentEvents();
        };
        TTResizableColumn.ctorParameters = function () { return [
            { type: TreeTable },
            { type: core.ElementRef },
            { type: core.NgZone }
        ]; };
        __decorate([
            core.Input()
        ], TTResizableColumn.prototype, "ttResizableColumnDisabled", void 0);
        TTResizableColumn = __decorate([
            core.Directive({
                selector: '[ttResizableColumn]'
            })
        ], TTResizableColumn);
        return TTResizableColumn;
    }());
    var TTReorderableColumn = /** @class */ (function () {
        function TTReorderableColumn(tt, el, zone) {
            this.tt = tt;
            this.el = el;
            this.zone = zone;
        }
        TTReorderableColumn.prototype.ngAfterViewInit = function () {
            if (this.isEnabled()) {
                this.bindEvents();
            }
        };
        TTReorderableColumn.prototype.bindEvents = function () {
            var _this = this;
            this.zone.runOutsideAngular(function () {
                _this.mouseDownListener = _this.onMouseDown.bind(_this);
                _this.el.nativeElement.addEventListener('mousedown', _this.mouseDownListener);
                _this.dragStartListener = _this.onDragStart.bind(_this);
                _this.el.nativeElement.addEventListener('dragstart', _this.dragStartListener);
                _this.dragOverListener = _this.onDragEnter.bind(_this);
                _this.el.nativeElement.addEventListener('dragover', _this.dragOverListener);
                _this.dragEnterListener = _this.onDragEnter.bind(_this);
                _this.el.nativeElement.addEventListener('dragenter', _this.dragEnterListener);
                _this.dragLeaveListener = _this.onDragLeave.bind(_this);
                _this.el.nativeElement.addEventListener('dragleave', _this.dragLeaveListener);
            });
        };
        TTReorderableColumn.prototype.unbindEvents = function () {
            if (this.mouseDownListener) {
                document.removeEventListener('mousedown', this.mouseDownListener);
                this.mouseDownListener = null;
            }
            if (this.dragOverListener) {
                document.removeEventListener('dragover', this.dragOverListener);
                this.dragOverListener = null;
            }
            if (this.dragEnterListener) {
                document.removeEventListener('dragenter', this.dragEnterListener);
                this.dragEnterListener = null;
            }
            if (this.dragEnterListener) {
                document.removeEventListener('dragenter', this.dragEnterListener);
                this.dragEnterListener = null;
            }
            if (this.dragLeaveListener) {
                document.removeEventListener('dragleave', this.dragLeaveListener);
                this.dragLeaveListener = null;
            }
        };
        TTReorderableColumn.prototype.onMouseDown = function (event) {
            if (event.target.nodeName === 'INPUT' || dom.DomHandler.hasClass(event.target, 'ui-column-resizer'))
                this.el.nativeElement.draggable = false;
            else
                this.el.nativeElement.draggable = true;
        };
        TTReorderableColumn.prototype.onDragStart = function (event) {
            this.tt.onColumnDragStart(event, this.el.nativeElement);
        };
        TTReorderableColumn.prototype.onDragOver = function (event) {
            event.preventDefault();
        };
        TTReorderableColumn.prototype.onDragEnter = function (event) {
            this.tt.onColumnDragEnter(event, this.el.nativeElement);
        };
        TTReorderableColumn.prototype.onDragLeave = function (event) {
            this.tt.onColumnDragLeave(event);
        };
        TTReorderableColumn.prototype.onDrop = function (event) {
            if (this.isEnabled()) {
                this.tt.onColumnDrop(event, this.el.nativeElement);
            }
        };
        TTReorderableColumn.prototype.isEnabled = function () {
            return this.ttReorderableColumnDisabled !== true;
        };
        TTReorderableColumn.prototype.ngOnDestroy = function () {
            this.unbindEvents();
        };
        TTReorderableColumn.ctorParameters = function () { return [
            { type: TreeTable },
            { type: core.ElementRef },
            { type: core.NgZone }
        ]; };
        __decorate([
            core.Input()
        ], TTReorderableColumn.prototype, "ttReorderableColumnDisabled", void 0);
        __decorate([
            core.HostListener('drop', ['$event'])
        ], TTReorderableColumn.prototype, "onDrop", null);
        TTReorderableColumn = __decorate([
            core.Directive({
                selector: '[ttReorderableColumn]'
            })
        ], TTReorderableColumn);
        return TTReorderableColumn;
    }());
    var TTSelectableRow = /** @class */ (function () {
        function TTSelectableRow(tt, tableService) {
            var _this = this;
            this.tt = tt;
            this.tableService = tableService;
            if (this.isEnabled()) {
                this.subscription = this.tt.tableService.selectionSource$.subscribe(function () {
                    _this.selected = _this.tt.isSelected(_this.rowNode.node);
                });
            }
        }
        TTSelectableRow.prototype.ngOnInit = function () {
            if (this.isEnabled()) {
                this.selected = this.tt.isSelected(this.rowNode.node);
            }
        };
        TTSelectableRow.prototype.onClick = function (event) {
            if (this.isEnabled()) {
                this.tt.handleRowClick({
                    originalEvent: event,
                    rowNode: this.rowNode
                });
            }
        };
        TTSelectableRow.prototype.onEnterKey = function (event) {
            if (event.which === 13) {
                this.onClick(event);
            }
        };
        TTSelectableRow.prototype.onTouchEnd = function (event) {
            if (this.isEnabled()) {
                this.tt.handleRowTouchEnd(event);
            }
        };
        TTSelectableRow.prototype.isEnabled = function () {
            return this.ttSelectableRowDisabled !== true;
        };
        TTSelectableRow.prototype.ngOnDestroy = function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        TTSelectableRow.ctorParameters = function () { return [
            { type: TreeTable },
            { type: TreeTableService }
        ]; };
        __decorate([
            core.Input("ttSelectableRow")
        ], TTSelectableRow.prototype, "rowNode", void 0);
        __decorate([
            core.Input()
        ], TTSelectableRow.prototype, "ttSelectableRowDisabled", void 0);
        __decorate([
            core.HostListener('click', ['$event'])
        ], TTSelectableRow.prototype, "onClick", null);
        __decorate([
            core.HostListener('keydown', ['$event'])
        ], TTSelectableRow.prototype, "onEnterKey", null);
        __decorate([
            core.HostListener('touchend', ['$event'])
        ], TTSelectableRow.prototype, "onTouchEnd", null);
        TTSelectableRow = __decorate([
            core.Directive({
                selector: '[ttSelectableRow]',
                host: {
                    '[class.ui-state-highlight]': 'selected'
                }
            })
        ], TTSelectableRow);
        return TTSelectableRow;
    }());
    var TTSelectableRowDblClick = /** @class */ (function () {
        function TTSelectableRowDblClick(tt, tableService) {
            var _this = this;
            this.tt = tt;
            this.tableService = tableService;
            if (this.isEnabled()) {
                this.subscription = this.tt.tableService.selectionSource$.subscribe(function () {
                    _this.selected = _this.tt.isSelected(_this.rowNode.node);
                });
            }
        }
        TTSelectableRowDblClick.prototype.ngOnInit = function () {
            if (this.isEnabled()) {
                this.selected = this.tt.isSelected(this.rowNode.node);
            }
        };
        TTSelectableRowDblClick.prototype.onClick = function (event) {
            if (this.isEnabled()) {
                this.tt.handleRowClick({
                    originalEvent: event,
                    rowNode: this.rowNode
                });
            }
        };
        TTSelectableRowDblClick.prototype.isEnabled = function () {
            return this.ttSelectableRowDisabled !== true;
        };
        TTSelectableRowDblClick.prototype.ngOnDestroy = function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        TTSelectableRowDblClick.ctorParameters = function () { return [
            { type: TreeTable },
            { type: TreeTableService }
        ]; };
        __decorate([
            core.Input("ttSelectableRowDblClick")
        ], TTSelectableRowDblClick.prototype, "rowNode", void 0);
        __decorate([
            core.Input()
        ], TTSelectableRowDblClick.prototype, "ttSelectableRowDisabled", void 0);
        __decorate([
            core.HostListener('dblclick', ['$event'])
        ], TTSelectableRowDblClick.prototype, "onClick", null);
        TTSelectableRowDblClick = __decorate([
            core.Directive({
                selector: '[ttSelectableRowDblClick]',
                host: {
                    '[class.ui-state-highlight]': 'selected'
                }
            })
        ], TTSelectableRowDblClick);
        return TTSelectableRowDblClick;
    }());
    var TTContextMenuRow = /** @class */ (function () {
        function TTContextMenuRow(tt, tableService, el) {
            var _this = this;
            this.tt = tt;
            this.tableService = tableService;
            this.el = el;
            if (this.isEnabled()) {
                this.subscription = this.tt.tableService.contextMenuSource$.subscribe(function (node) {
                    _this.selected = _this.tt.equals(_this.rowNode.node, node);
                });
            }
        }
        TTContextMenuRow.prototype.onContextMenu = function (event) {
            if (this.isEnabled()) {
                this.tt.handleRowRightClick({
                    originalEvent: event,
                    rowNode: this.rowNode
                });
                this.el.nativeElement.focus();
                event.preventDefault();
            }
        };
        TTContextMenuRow.prototype.isEnabled = function () {
            return this.ttContextMenuRowDisabled !== true;
        };
        TTContextMenuRow.prototype.ngOnDestroy = function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        TTContextMenuRow.ctorParameters = function () { return [
            { type: TreeTable },
            { type: TreeTableService },
            { type: core.ElementRef }
        ]; };
        __decorate([
            core.Input("ttContextMenuRow")
        ], TTContextMenuRow.prototype, "rowNode", void 0);
        __decorate([
            core.Input()
        ], TTContextMenuRow.prototype, "ttContextMenuRowDisabled", void 0);
        __decorate([
            core.HostListener('contextmenu', ['$event'])
        ], TTContextMenuRow.prototype, "onContextMenu", null);
        TTContextMenuRow = __decorate([
            core.Directive({
                selector: '[ttContextMenuRow]',
                host: {
                    '[class.ui-contextmenu-selected]': 'selected',
                    '[attr.tabindex]': 'isEnabled() ? 0 : undefined'
                }
            })
        ], TTContextMenuRow);
        return TTContextMenuRow;
    }());
    var TTCheckbox = /** @class */ (function () {
        function TTCheckbox(tt, tableService) {
            var _this = this;
            this.tt = tt;
            this.tableService = tableService;
            this.subscription = this.tt.tableService.selectionSource$.subscribe(function () {
                _this.checked = _this.tt.isSelected(_this.rowNode.node);
            });
        }
        TTCheckbox.prototype.ngOnInit = function () {
            this.checked = this.tt.isSelected(this.rowNode.node);
        };
        TTCheckbox.prototype.onClick = function (event) {
            if (!this.disabled) {
                this.tt.toggleNodeWithCheckbox({
                    originalEvent: event,
                    rowNode: this.rowNode
                });
            }
            dom.DomHandler.clearSelection();
        };
        TTCheckbox.prototype.onFocus = function () {
            dom.DomHandler.addClass(this.boxViewChild.nativeElement, 'ui-state-focus');
        };
        TTCheckbox.prototype.onBlur = function () {
            dom.DomHandler.removeClass(this.boxViewChild.nativeElement, 'ui-state-focus');
        };
        TTCheckbox.prototype.ngOnDestroy = function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        TTCheckbox.ctorParameters = function () { return [
            { type: TreeTable },
            { type: TreeTableService }
        ]; };
        __decorate([
            core.Input()
        ], TTCheckbox.prototype, "disabled", void 0);
        __decorate([
            core.Input("value")
        ], TTCheckbox.prototype, "rowNode", void 0);
        __decorate([
            core.ViewChild('box')
        ], TTCheckbox.prototype, "boxViewChild", void 0);
        TTCheckbox = __decorate([
            core.Component({
                selector: 'p-treeTableCheckbox',
                template: "\n        <div class=\"ui-chkbox ui-treetable-chkbox ui-widget\" (click)=\"onClick($event)\">\n            <div class=\"ui-helper-hidden-accessible\">\n                <input type=\"checkbox\" [checked]=\"checked\" (focus)=\"onFocus()\" (blur)=\"onBlur()\">\n            </div>\n            <div #box [ngClass]=\"{'ui-chkbox-box ui-widget ui-state-default':true,\n                'ui-state-active':checked, 'ui-state-disabled':disabled}\"  role=\"checkbox\" [attr.aria-checked]=\"checked\">\n                <span class=\"ui-chkbox-icon ui-clickable pi\" [ngClass]=\"{'pi-check':checked, 'pi-minus': rowNode.node.partialSelected}\"></span>\n            </div>\n        </div>\n    "
            })
        ], TTCheckbox);
        return TTCheckbox;
    }());
    var TTHeaderCheckbox = /** @class */ (function () {
        function TTHeaderCheckbox(tt, tableService) {
            var _this = this;
            this.tt = tt;
            this.tableService = tableService;
            this.valueChangeSubscription = this.tt.tableService.uiUpdateSource$.subscribe(function () {
                _this.checked = _this.updateCheckedState();
            });
            this.selectionChangeSubscription = this.tt.tableService.selectionSource$.subscribe(function () {
                _this.checked = _this.updateCheckedState();
            });
        }
        TTHeaderCheckbox.prototype.ngOnInit = function () {
            this.checked = this.updateCheckedState();
        };
        TTHeaderCheckbox.prototype.onClick = function (event, checked) {
            if (this.tt.value && this.tt.value.length > 0) {
                this.tt.toggleNodesWithCheckbox(event, !checked);
            }
            dom.DomHandler.clearSelection();
        };
        TTHeaderCheckbox.prototype.onFocus = function () {
            dom.DomHandler.addClass(this.boxViewChild.nativeElement, 'ui-state-focus');
        };
        TTHeaderCheckbox.prototype.onBlur = function () {
            dom.DomHandler.removeClass(this.boxViewChild.nativeElement, 'ui-state-focus');
        };
        TTHeaderCheckbox.prototype.ngOnDestroy = function () {
            if (this.selectionChangeSubscription) {
                this.selectionChangeSubscription.unsubscribe();
            }
            if (this.valueChangeSubscription) {
                this.valueChangeSubscription.unsubscribe();
            }
        };
        TTHeaderCheckbox.prototype.updateCheckedState = function () {
            var e_10, _a;
            var checked;
            var data = this.tt.filteredNodes || this.tt.value;
            if (data) {
                try {
                    for (var data_2 = __values(data), data_2_1 = data_2.next(); !data_2_1.done; data_2_1 = data_2.next()) {
                        var node = data_2_1.value;
                        if (this.tt.isSelected(node)) {
                            checked = true;
                        }
                        else {
                            checked = false;
                            break;
                        }
                    }
                }
                catch (e_10_1) { e_10 = { error: e_10_1 }; }
                finally {
                    try {
                        if (data_2_1 && !data_2_1.done && (_a = data_2.return)) _a.call(data_2);
                    }
                    finally { if (e_10) throw e_10.error; }
                }
            }
            else {
                checked = false;
            }
            return checked;
        };
        TTHeaderCheckbox.ctorParameters = function () { return [
            { type: TreeTable },
            { type: TreeTableService }
        ]; };
        __decorate([
            core.ViewChild('box')
        ], TTHeaderCheckbox.prototype, "boxViewChild", void 0);
        TTHeaderCheckbox = __decorate([
            core.Component({
                selector: 'p-treeTableHeaderCheckbox',
                template: "\n        <div class=\"ui-chkbox ui-treetable-header-chkbox ui-widget\" (click)=\"onClick($event, cb.checked)\">\n            <div class=\"ui-helper-hidden-accessible\">\n                <input #cb type=\"checkbox\" [checked]=\"checked\" (focus)=\"onFocus()\" (blur)=\"onBlur()\" [disabled]=\"!tt.value||tt.value.length === 0\">\n            </div>\n            <div #box [ngClass]=\"{'ui-chkbox-box ui-widget ui-state-default':true,\n                'ui-state-active':checked, 'ui-state-disabled': (!tt.value || tt.value.length === 0)}\"  role=\"checkbox\" [attr.aria-checked]=\"checked\">\n                <span class=\"ui-chkbox-icon ui-clickable\" [ngClass]=\"{'pi pi-check':checked}\"></span>\n            </div>\n        </div>\n    "
            })
        ], TTHeaderCheckbox);
        return TTHeaderCheckbox;
    }());
    var TTEditableColumn = /** @class */ (function () {
        function TTEditableColumn(tt, el, zone) {
            this.tt = tt;
            this.el = el;
            this.zone = zone;
        }
        TTEditableColumn.prototype.ngAfterViewInit = function () {
            if (this.isEnabled()) {
                dom.DomHandler.addClass(this.el.nativeElement, 'ui-editable-column');
            }
        };
        TTEditableColumn.prototype.onClick = function (event) {
            if (this.isEnabled()) {
                this.tt.editingCellClick = true;
                if (this.tt.editingCell) {
                    if (this.tt.editingCell !== this.el.nativeElement) {
                        if (!this.tt.isEditingCellValid()) {
                            return;
                        }
                        dom.DomHandler.removeClass(this.tt.editingCell, 'ui-editing-cell');
                        this.openCell();
                    }
                }
                else {
                    this.openCell();
                }
            }
        };
        TTEditableColumn.prototype.openCell = function () {
            var _this = this;
            this.tt.updateEditingCell(this.el.nativeElement);
            dom.DomHandler.addClass(this.el.nativeElement, 'ui-editing-cell');
            this.tt.onEditInit.emit({ field: this.field, data: this.data });
            this.zone.runOutsideAngular(function () {
                setTimeout(function () {
                    var focusable = dom.DomHandler.findSingle(_this.el.nativeElement, 'input, textarea');
                    if (focusable) {
                        focusable.focus();
                    }
                }, 50);
            });
        };
        TTEditableColumn.prototype.closeEditingCell = function () {
            dom.DomHandler.removeClass(this.tt.editingCell, 'ui-editing-cell');
            this.tt.editingCell = null;
            this.tt.unbindDocumentEditListener();
        };
        TTEditableColumn.prototype.onKeyDown = function (event) {
            if (this.isEnabled()) {
                //enter
                if (event.keyCode == 13) {
                    if (this.tt.isEditingCellValid()) {
                        dom.DomHandler.removeClass(this.tt.editingCell, 'ui-editing-cell');
                        this.closeEditingCell();
                        this.tt.onEditComplete.emit({ field: this.field, data: this.data });
                    }
                    event.preventDefault();
                }
                //escape
                else if (event.keyCode == 27) {
                    if (this.tt.isEditingCellValid()) {
                        dom.DomHandler.removeClass(this.tt.editingCell, 'ui-editing-cell');
                        this.closeEditingCell();
                        this.tt.onEditCancel.emit({ field: this.field, data: this.data });
                    }
                    event.preventDefault();
                }
                //tab
                else if (event.keyCode == 9) {
                    this.tt.onEditComplete.emit({ field: this.field, data: this.data });
                    if (event.shiftKey)
                        this.moveToPreviousCell(event);
                    else
                        this.moveToNextCell(event);
                }
            }
        };
        TTEditableColumn.prototype.findCell = function (element) {
            if (element) {
                var cell = element;
                while (cell && !dom.DomHandler.hasClass(cell, 'ui-editing-cell')) {
                    cell = cell.parentElement;
                }
                return cell;
            }
            else {
                return null;
            }
        };
        TTEditableColumn.prototype.moveToPreviousCell = function (event) {
            var currentCell = this.findCell(event.target);
            var row = currentCell.parentElement;
            var targetCell = this.findPreviousEditableColumn(currentCell);
            if (targetCell) {
                dom.DomHandler.invokeElementMethod(targetCell, 'click');
                event.preventDefault();
            }
        };
        TTEditableColumn.prototype.moveToNextCell = function (event) {
            var currentCell = this.findCell(event.target);
            var row = currentCell.parentElement;
            var targetCell = this.findNextEditableColumn(currentCell);
            if (targetCell) {
                dom.DomHandler.invokeElementMethod(targetCell, 'click');
                event.preventDefault();
            }
        };
        TTEditableColumn.prototype.findPreviousEditableColumn = function (cell) {
            var prevCell = cell.previousElementSibling;
            if (!prevCell) {
                var previousRow = cell.parentElement ? cell.parentElement.previousElementSibling : null;
                if (previousRow) {
                    prevCell = previousRow.lastElementChild;
                }
            }
            if (prevCell) {
                if (dom.DomHandler.hasClass(prevCell, 'ui-editable-column'))
                    return prevCell;
                else
                    return this.findPreviousEditableColumn(prevCell);
            }
            else {
                return null;
            }
        };
        TTEditableColumn.prototype.findNextEditableColumn = function (cell) {
            var nextCell = cell.nextElementSibling;
            if (!nextCell) {
                var nextRow = cell.parentElement ? cell.parentElement.nextElementSibling : null;
                if (nextRow) {
                    nextCell = nextRow.firstElementChild;
                }
            }
            if (nextCell) {
                if (dom.DomHandler.hasClass(nextCell, 'ui-editable-column'))
                    return nextCell;
                else
                    return this.findNextEditableColumn(nextCell);
            }
            else {
                return null;
            }
        };
        TTEditableColumn.prototype.isEnabled = function () {
            return this.ttEditableColumnDisabled !== true;
        };
        TTEditableColumn.ctorParameters = function () { return [
            { type: TreeTable },
            { type: core.ElementRef },
            { type: core.NgZone }
        ]; };
        __decorate([
            core.Input("ttEditableColumn")
        ], TTEditableColumn.prototype, "data", void 0);
        __decorate([
            core.Input("ttEditableColumnField")
        ], TTEditableColumn.prototype, "field", void 0);
        __decorate([
            core.Input()
        ], TTEditableColumn.prototype, "ttEditableColumnDisabled", void 0);
        __decorate([
            core.HostListener('click', ['$event'])
        ], TTEditableColumn.prototype, "onClick", null);
        __decorate([
            core.HostListener('keydown', ['$event'])
        ], TTEditableColumn.prototype, "onKeyDown", null);
        TTEditableColumn = __decorate([
            core.Directive({
                selector: '[ttEditableColumn]'
            })
        ], TTEditableColumn);
        return TTEditableColumn;
    }());
    var TreeTableCellEditor = /** @class */ (function () {
        function TreeTableCellEditor(tt, editableColumn) {
            this.tt = tt;
            this.editableColumn = editableColumn;
        }
        TreeTableCellEditor.prototype.ngAfterContentInit = function () {
            var _this = this;
            this.templates.forEach(function (item) {
                switch (item.getType()) {
                    case 'input':
                        _this.inputTemplate = item.template;
                        break;
                    case 'output':
                        _this.outputTemplate = item.template;
                        break;
                }
            });
        };
        TreeTableCellEditor.ctorParameters = function () { return [
            { type: TreeTable },
            { type: TTEditableColumn }
        ]; };
        __decorate([
            core.ContentChildren(api.PrimeTemplate)
        ], TreeTableCellEditor.prototype, "templates", void 0);
        TreeTableCellEditor = __decorate([
            core.Component({
                selector: 'p-treeTableCellEditor',
                template: "\n        <ng-container *ngIf=\"tt.editingCell === editableColumn.el.nativeElement\">\n            <ng-container *ngTemplateOutlet=\"inputTemplate\"></ng-container>\n        </ng-container>\n        <ng-container *ngIf=\"!tt.editingCell || tt.editingCell !== editableColumn.el.nativeElement\">\n            <ng-container *ngTemplateOutlet=\"outputTemplate\"></ng-container>\n        </ng-container>\n    "
            })
        ], TreeTableCellEditor);
        return TreeTableCellEditor;
    }());
    var TTRow = /** @class */ (function () {
        function TTRow(tt, el, zone) {
            this.tt = tt;
            this.el = el;
            this.zone = zone;
        }
        TTRow.prototype.onKeyDown = function (event) {
            switch (event.which) {
                //down arrow
                case 40:
                    var nextRow = this.el.nativeElement.nextElementSibling;
                    if (nextRow) {
                        nextRow.focus();
                    }
                    event.preventDefault();
                    break;
                //down arrow
                case 38:
                    var prevRow = this.el.nativeElement.previousElementSibling;
                    if (prevRow) {
                        prevRow.focus();
                    }
                    event.preventDefault();
                    break;
                //left arrow
                case 37:
                    if (this.rowNode.node.expanded) {
                        this.tt.toggleRowIndex = dom.DomHandler.index(this.el.nativeElement);
                        this.rowNode.node.expanded = false;
                        this.tt.onNodeCollapse.emit({
                            originalEvent: event,
                            node: this.rowNode.node
                        });
                        this.tt.updateSerializedValue();
                        this.tt.tableService.onUIUpdate(this.tt.value);
                        this.restoreFocus();
                    }
                    break;
                //right arrow
                case 39:
                    if (!this.rowNode.node.expanded) {
                        this.tt.toggleRowIndex = dom.DomHandler.index(this.el.nativeElement);
                        this.rowNode.node.expanded = true;
                        this.tt.onNodeExpand.emit({
                            originalEvent: event,
                            node: this.rowNode.node
                        });
                        this.tt.updateSerializedValue();
                        this.tt.tableService.onUIUpdate(this.tt.value);
                        this.restoreFocus();
                    }
                    break;
            }
        };
        TTRow.prototype.restoreFocus = function () {
            var _this = this;
            this.zone.runOutsideAngular(function () {
                setTimeout(function () {
                    var row = dom.DomHandler.findSingle(_this.tt.containerViewChild.nativeElement, '.ui-treetable-tbody').children[_this.tt.toggleRowIndex];
                    if (row) {
                        row.focus();
                    }
                }, 25);
            });
        };
        TTRow.ctorParameters = function () { return [
            { type: TreeTable },
            { type: core.ElementRef },
            { type: core.NgZone }
        ]; };
        __decorate([
            core.Input('ttRow')
        ], TTRow.prototype, "rowNode", void 0);
        __decorate([
            core.HostListener('keydown', ['$event'])
        ], TTRow.prototype, "onKeyDown", null);
        TTRow = __decorate([
            core.Directive({
                selector: '[ttRow]',
                host: {
                    '[attr.tabindex]': '"0"'
                }
            })
        ], TTRow);
        return TTRow;
    }());
    var TreeTableToggler = /** @class */ (function () {
        function TreeTableToggler(tt) {
            this.tt = tt;
        }
        TreeTableToggler.prototype.onClick = function (event) {
            this.rowNode.node.expanded = !this.rowNode.node.expanded;
            if (this.rowNode.node.expanded) {
                this.tt.onNodeExpand.emit({
                    originalEvent: event,
                    node: this.rowNode.node
                });
            }
            else {
                this.tt.onNodeCollapse.emit({
                    originalEvent: event,
                    node: this.rowNode.node
                });
            }
            this.tt.updateSerializedValue();
            this.tt.tableService.onUIUpdate(this.tt.value);
            event.preventDefault();
        };
        TreeTableToggler.ctorParameters = function () { return [
            { type: TreeTable }
        ]; };
        __decorate([
            core.Input()
        ], TreeTableToggler.prototype, "rowNode", void 0);
        TreeTableToggler = __decorate([
            core.Component({
                selector: 'p-treeTableToggler',
                template: "\n        <a class=\"ui-treetable-toggler ui-unselectable-text\" (click)=\"onClick($event)\"\n            [style.visibility]=\"rowNode.node.leaf === false || (rowNode.node.children && rowNode.node.children.length) ? 'visible' : 'hidden'\" [style.marginLeft]=\"rowNode.level * 16 + 'px'\">\n            <i [ngClass]=\"rowNode.node.expanded ? 'pi pi-fw pi-chevron-down' : 'pi pi-fw pi-chevron-right'\"></i>\n        </a>\n    "
            })
        ], TreeTableToggler);
        return TreeTableToggler;
    }());
    var TreeTableModule = /** @class */ (function () {
        function TreeTableModule() {
        }
        TreeTableModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, paginator.PaginatorModule, scrolling.ScrollingModule],
                exports: [TreeTable, api.SharedModule, TreeTableToggler, TTSortableColumn, TTSortIcon, TTResizableColumn, TTRow, TTReorderableColumn, TTSelectableRow, TTSelectableRowDblClick, TTContextMenuRow, TTCheckbox, TTHeaderCheckbox, TTEditableColumn, TreeTableCellEditor, scrolling.ScrollingModule],
                declarations: [TreeTable, TreeTableToggler, TTScrollableView, TTBody, TTSortableColumn, TTSortIcon, TTResizableColumn, TTRow, TTReorderableColumn, TTSelectableRow, TTSelectableRowDblClick, TTContextMenuRow, TTCheckbox, TTHeaderCheckbox, TTEditableColumn, TreeTableCellEditor]
            })
        ], TreeTableModule);
        return TreeTableModule;
    }());

    exports.TTBody = TTBody;
    exports.TTCheckbox = TTCheckbox;
    exports.TTContextMenuRow = TTContextMenuRow;
    exports.TTEditableColumn = TTEditableColumn;
    exports.TTHeaderCheckbox = TTHeaderCheckbox;
    exports.TTReorderableColumn = TTReorderableColumn;
    exports.TTResizableColumn = TTResizableColumn;
    exports.TTRow = TTRow;
    exports.TTScrollableView = TTScrollableView;
    exports.TTSelectableRow = TTSelectableRow;
    exports.TTSelectableRowDblClick = TTSelectableRowDblClick;
    exports.TTSortIcon = TTSortIcon;
    exports.TTSortableColumn = TTSortableColumn;
    exports.TreeTable = TreeTable;
    exports.TreeTableCellEditor = TreeTableCellEditor;
    exports.TreeTableModule = TreeTableModule;
    exports.TreeTableService = TreeTableService;
    exports.TreeTableToggler = TreeTableToggler;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
